// @canva/app-ui-kit/dist/esm/ui/base/divider/private/divider.js
import { jsx as _jsx } from "react/jsx-runtime";
import * as stylex from '../../../../platforms/libs/stylex/stylex';
import 'react';
import { easelMargin } from '../../style_helpers/easel_margin';
import '../../tokens/primitive_tokens.stylex';
import '../../tokens/semantic_tokens.stylex';
export const Divider = ({ role, direction, borderWidth, inset = 'none', spacing = 'none' })=>{
    const isVertical = direction === 'vertical';
    return _jsx("hr", {
        role: role,
        "aria-orientation": role !== 'none' && isVertical ? 'vertical' : undefined,
        ...stylex.props([
            styles.root,
            isVertical && styles.vertical,
            borderWidth === '200' && (isVertical ? styles.borderWidth200Vertical : styles.borderWidth200),
            inset !== 'none' && (isVertical ? [
                insetStyles[inset],
                verticalInsetHeightStyles[inset]
            ] : easelMargin.inline[inset]),
            spacing !== 'none' && (isVertical ? easelMargin.inline[spacing] : spacingStyles[spacing])
        ])
    });
};
const styles = {
    root: {
        kqGvvJ: "x10im51j",
        kUOVxO: "xrxpjvj",
        k8ry5P: "x18oe1m7",
        ke9TFa: "x10w94by",
        kVhnKS: "xstzfhl",
        k2ei4v: "x14e42zd",
        kB7OPa: "x9f619",
        kL6WhQ: "xwxg45u",
        kfdmCh: "x1q0q8m5",
        kt9PQ7: "xso031l",
        kPef9Z: "x1ejq31n",
        kEafiO: "x972fbf",
        kZKoxP: "xjm9jq1",
        ["$$css"]: true
    },
    vertical: {
        kBCPoo: "xvjrof5",
        k8ry5P: "x18b5jzi",
        ke9TFa: "x1lun4ml",
        k1xSpc: "x1rg5ohu",
        kfdmCh: "x1sy0etr",
        kt9PQ7: "x1qhh985",
        kZKoxP: "x5yr21d",
        kzqmXN: "x1i1rx1s",
        ["$$css"]: true
    },
    borderWidth200: {
        kt9PQ7: "xlxy82",
        kZKoxP: "x36qwtl",
        ["$$css"]: true
    },
    borderWidth200Vertical: {
        ke9TFa: "x1alpsbp",
        kzqmXN: "xfo62xy",
        ["$$css"]: true
    }
};
const verticalInsetHeightStyles = {
    "0.5u": {
        kZKoxP: "xzkfkfq",
        ["$$css"]: true
    },
    "1u": {
        kZKoxP: "x1lnd9bg",
        ["$$css"]: true
    },
    "1.5u": {
        kZKoxP: "xlqzx8c",
        ["$$css"]: true
    },
    "2u": {
        kZKoxP: "xwivwg1",
        ["$$css"]: true
    }
};
const insetStyles = {
    "0.5u": {
        kqGvvJ: "xm6m106",
        ["$$css"]: true
    },
    "1u": {
        kqGvvJ: "x1dykhsd",
        ["$$css"]: true
    },
    "1.5u": {
        kqGvvJ: "xgngeza",
        ["$$css"]: true
    },
    "2u": {
        kqGvvJ: "x19d84vq",
        ["$$css"]: true
    }
};
const spacingStyles = {
    "0.25u": {
        kqGvvJ: "x1dqyc2m",
        ["$$css"]: true
    },
    "0.5u": {
        kqGvvJ: "xm6m106",
        ["$$css"]: true
    },
    "1u": {
        kqGvvJ: "x1dykhsd",
        ["$$css"]: true
    },
    "1.5u": {
        kqGvvJ: "xgngeza",
        ["$$css"]: true
    },
    "2u": {
        kqGvvJ: "x19d84vq",
        ["$$css"]: true
    }
};
