// @canva/app-ui-kit/dist/cjs/ui/apps/developing/ui_kit/components/card/video/video.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "VideoCard", {
    enumerable: true,
    get: function() {
        return VideoCard;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _preconditions = require('../../../../../../../base/preconditions');
const _classnames = _interop_require_default(require("classnames"));
const _react = _interop_require_wildcard(require("react"));
const _box = require('../../../../../../base/box/box');
const _card = require('../../../../../../base/card/card');
const _duration = require('../../../../../../base/date_time/duration/duration');
const _icon = require('../../../../../../base/icons/animated_spinner/icon');
const _badge = require("../../badge/badge");
const _draggable = require("../draggable/draggable");
const _videocss = _interop_require_default(require("./video.css"));
const _videomessages = require("./video.messages");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const VideoCard = (props)=>(0, _jsxruntime.jsx)(_draggable.Draggable, {
        ...props,
        children: (conf)=>(0, _jsxruntime.jsx)(InternalVideoCard, {
                ...props,
                ...conf
            })
    });
const InternalVideoCard = (props)=>{
    const { ariaLabel, thumbnailUrl, mimeType, onDragStart, onDragEnd, onClick, borderRadius = 'none', thumbnailHeight, thumbnailAspectRatio, bottomEnd, bottomEndVisibility, selectable, selected, disabled, loading, onVideoLoad, onImageLoad } = props;
    const ref = isVideoProps(props) ? props.ref : undefined;
    const video = _react.useRef(null);
    const setVideoRef = _react.useCallback((node)=>{
        if (node != null) video.current = node;
    }, []);
    const [status, setStatus] = _react.useState('uncontrolled');
    _react.useImperativeHandle(ref, ()=>({
            play: ()=>{
                setStatus('playing');
                video.current?.play();
            },
            pause: ()=>{
                setStatus('paused');
                video.current?.pause();
            },
            restart: ()=>{
                setStatus('playing');
                if (video.current) {
                    video.current.currentTime = 0;
                    video.current.play();
                }
            },
            isPlaying: ()=>status === 'playing',
            isPaused: ()=>status === 'paused'
        }), [
        status
    ]);
    _react.useEffect(()=>{
        const dispose = ()=>video.current?.pause();
        return dispose;
    }, []);
    const whenToPlay = status === 'uncontrolled' ? 'on-hover' : 'now';
    const autoplay = status !== 'paused';
    const commonProps = {
        buttonAriaLabel: ariaLabel,
        onClick,
        draggable: onDragStart != null,
        hoverEffect: 'none',
        thumbnailHeight,
        selectableMode: selectable ? 'primary' : 'none',
        selected,
        disabled,
        onDragStart,
        onDragEnd
    };
    if (isVideoProps(props)) {
        const badgeLabel = getBadgeLabel(mimeType, props.durationInSeconds);
        return (0, _jsxruntime.jsx)(_card.Card, {
            decorators: (0, _jsxruntime.jsxs)(_jsxruntime.Fragment, {
                children: [
                    (0, _jsxruntime.jsx)(_card.CardDecorator, {
                        location: "bottom-start",
                        children: (0, _jsxruntime.jsx)(_badge.Badge, {
                            tone: "contrast",
                            text: badgeLabel
                        })
                    }),
                    bottomEnd && (0, _jsxruntime.jsx)(_card.CardDecorator, {
                        location: "bottom-end",
                        visibility: bottomEndVisibility,
                        isInteractive: true,
                        children: bottomEnd
                    })
                ]
            }),
            ...commonProps,
            thumbnail: (0, _jsxruntime.jsxs)(_card.CardThumbnailContainer, {
                background: "none",
                padding: "none",
                borderRadius: borderRadius === 'standard' ? 'element' : 'none',
                children: [
                    (0, _jsxruntime.jsx)(_card.CardVideoThumbnail, {
                        ref: setVideoRef,
                        imageSrc: thumbnailUrl,
                        videoSrc: props.videoPreviewUrl,
                        whenToPlay: whenToPlay,
                        autoplay: autoplay,
                        borderRadius: borderRadius === 'standard' ? 'element' : 'none',
                        height: thumbnailHeight,
                        onVideoLoad: onVideoLoad,
                        onImageLoad: onImageLoad,
                        className: (0, _classnames.default)({
                            [_videocss.default.thumbnailAutoWidth]: !thumbnailAspectRatio && thumbnailHeight
                        })
                    }),
                    loading && (0, _jsxruntime.jsx)(_box.Box, {
                        className: _videocss.default.loading,
                        borderRadius: borderRadius === 'standard' ? 'legacyLarge' : 'none',
                        display: "flex",
                        justifyContent: "center",
                        alignItems: "center",
                        width: "full",
                        height: "full",
                        children: (0, _jsxruntime.jsx)(_icon.AnimatedSpinnerIcon, {
                            size: "medium",
                            className: _videocss.default.spinner
                        })
                    })
                ]
            }),
            hoverEffect: "scaleThumbnail",
            layout: "vertical",
            hoverBoundary: "thumbnail",
            borderRadius: "element"
        });
    }
    return (0, _jsxruntime.jsx)(_card.Card, {
        decorators: (0, _jsxruntime.jsxs)(_jsxruntime.Fragment, {
            children: [
                (0, _jsxruntime.jsx)(_card.CardDecorator, {
                    location: "bottom-start",
                    children: (0, _jsxruntime.jsx)(_badge.Badge, {
                        tone: "contrast",
                        text: _videomessages.VideoCardMessages.gifLabel()
                    })
                }),
                bottomEnd && (0, _jsxruntime.jsx)(_card.CardDecorator, {
                    location: "bottom-end",
                    visibility: bottomEndVisibility,
                    isInteractive: true,
                    children: bottomEnd
                })
            ]
        }),
        ...commonProps,
        thumbnail: (0, _jsxruntime.jsxs)(_card.CardThumbnailContainer, {
            background: "none",
            padding: "none",
            borderRadius: borderRadius === 'standard' ? 'element' : 'none',
            children: [
                (0, _jsxruntime.jsx)(_card.CardImageThumbnail, {
                    src: thumbnailUrl,
                    alt: props.alt,
                    borderRadius: borderRadius === 'standard' ? 'element' : 'none',
                    height: thumbnailHeight,
                    onImageLoad: onImageLoad,
                    className: (0, _classnames.default)({
                        [_videocss.default.thumbnailAutoWidth]: !thumbnailAspectRatio && thumbnailHeight
                    })
                }),
                loading && (0, _jsxruntime.jsx)(_box.Box, {
                    className: _videocss.default.loading,
                    borderRadius: borderRadius === 'standard' ? 'legacyLarge' : 'none',
                    display: "flex",
                    justifyContent: "center",
                    alignItems: "center",
                    width: "full",
                    height: "full",
                    children: (0, _jsxruntime.jsx)(_icon.AnimatedSpinnerIcon, {
                        size: "medium",
                        className: _videocss.default.spinner
                    })
                })
            ]
        }),
        hoverEffect: "scaleThumbnail",
        layout: "vertical",
        hoverBoundary: "thumbnail",
        borderRadius: "element"
    });
};
const isVideoProps = (props)=>{
    return props.videoPreviewUrl !== undefined;
};
const getBadgeLabel = (mimeType, duration)=>{
    switch(mimeType){
        case 'video/x-msvideo':
        case 'video/avi':
        case 'video/x-m4v':
        case 'video/x-matroska':
        case 'video/quicktime':
        case 'video/mp4':
        case 'video/mpeg':
        case 'video/webm':
            return duration ? (0, _duration.formatDuration)(duration, 'short') : _videomessages.VideoCardMessages.defaultLabel();
        case 'image/gif':
            return _videomessages.VideoCardMessages.gifLabel();
        default:
            throw new _preconditions.UnreachableError(mimeType);
    }
};
