// @canva/app-ui-kit/dist/esm/ui/base/provider/private/provider.js
import { jsx as _jsx } from "react/jsx-runtime";
import { Preconditions } from '../../../../base/preconditions';
import * as React from 'react';
import { InteractionManager } from '../../a11y/interaction_manager/interaction_manager';
import { useIsomorphicLayoutEffect } from '../../hydration/use_isomorphic_layout_effect';
import styles from '../provider.css';
import { featureFlag } from '../../../flags/feature_flag';
const flags = {
  enableButtonLoadingScan: featureFlag("106f47d", false),
  enableModernisedRadius: featureFlag("990db009", false)
};
const flagClassNameMap = {
  enableModernisedRadius: stripGlobal(styles.enableEaselModernisedRadius)
};
const EaselContext = React.createContext(null);
const FlagsContext = React.createContext(flags);
export function EaselProvider({
  enableAnimations = true,
  direction,
  disableFocusTraps,
  enableUserSelection,
  theme,
  enableHighColorContrast,
  autoplayVideos,
  disableDialogBlur,
  enableButtonLoadingScan,
  expectedPixelsPerRem,
  children
}) {
  return _jsx(InteractionManager, {
    children: _jsx(EaselContext.Provider, {
      value: {
        enableAnimations,
        direction,
        disableFocusTraps,
        enableUserSelection,
        theme,
        enableHighColorContrast,
        autoplayVideos,
        disableDialogBlur,
        expectedPixelsPerRem
      },
      children: _jsx(FlagsContext.Provider, {
        value: mergeFlags(flags, {
          enableButtonLoadingScan
        }),
        children: children
      })
    })
  });
}
export function useEaselFlags() {
  return React.useContext(FlagsContext);
}
export function useEaselFlagsSideEffect() {
  useIsomorphicLayoutEffect(() => {
    const classNames = Object.entries(flagClassNameMap).filter(([name]) => flags[name]).map(([_, value]) => value);
    document.documentElement.classList.add(...classNames);
  }, []);
}
function useEaselContext() {
  return Preconditions.checkExists(React.useContext(EaselContext), 'Did you forget to render a provider?');
}
export function useEnableAnimations() {
  return useEaselContext().enableAnimations;
}
export function useDirection() {
  return useEaselContext().direction;
}
export function useDisableFocusTraps() {
  return useEaselContext().disableFocusTraps;
}
export function useEnableUserSelection() {
  return useEaselContext().enableUserSelection;
}
export function useTheme() {
  return useEaselContext().theme;
}
export function useEnableHighColorContrast() {
  return useEaselContext().enableHighColorContrast;
}
export function useAutoplayVideos() {
  return useEaselContext().autoplayVideos;
}
export function useDisableDialogBlur() {
  return useEaselContext().disableDialogBlur;
}
export function useEnableButtonLoadingScan() {
  return useEaselFlags()?.enableButtonLoadingScan ?? false;
}
export function useExpectedPixelsPerRem() {
  return useEaselContext().expectedPixelsPerRem;
}
function mergeFlags(flags, overrides) {
  return {
    ...flags,
    ...Object.fromEntries(Object.entries(overrides ?? {}).filter(([_, v]) => v !== undefined))
  };
}
function stripGlobal(className) {
  return className?.replace(/^\s*:global\(\.(.+?)\)\s*$/, '$1') ?? '';
}
