// @canva/app-ui-kit/dist/cjs/ui/base/layout/private/spacer/spacer.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "Spacer", {
    enumerable: true,
    get: function() {
        return Spacer;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _stylex = _interop_require_wildcard(require('../../../../../platforms/libs/stylex/stylex'));
require("react");
const _metrics = require('../../../metrics/metrics');
const _responsive = require('../../../responsive/responsive');
const _spacertestids = require("./spacer.testids");
require("./vars.stylex");
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const Spacer = ({ size, direction = 'both' })=>(0, _jsxruntime.jsx)("span", {
        ..._stylex.props([
            styles.root,
            direction === 'horizontal' && styles.horizontal,
            direction === 'vertical' && styles.vertical,
            (0, _responsive.mapResponsiveValueToStyleX)(size, _metrics.getSpaceValue, styles.size, styles.sizeResponsive)
        ]),
        "aria-hidden": true,
        "data-testid": _spacertestids.SpacerTestIds.spacer
    });
const styles = {
    root: {
        k1xSpc: "x1lliihq",
        kmuXW: "x2lah0s",
        kZKoxP: "x11lv39n",
        kzqmXN: "x1ca76lx",
        ["$$css"]: true
    },
    horizontal: {
        k1xSpc: "x1rg5ohu",
        kZKoxP: "xt7dq6l",
        kAzted: "x4zw9a1",
        ["$$css"]: true
    },
    vertical: {
        k7Eaqz: "xgc0pbm",
        kzqmXN: "x14atkfc",
        ["$$css"]: true
    },
    size: (size)=>[
            {
                "--xqvgnxq": size != null ? "x1lahtpn" : size,
                ["$$css"]: true
            },
            {
                "--x---xqvgnxq": size != null ? size : undefined
            }
        ],
    sizeResponsive: (size)=>[
            {
                "--xqvgnxq": (size.default != null ? "xtlhp39 " : size.default) + "x4h0bxt " + "x1f3rlcv " + "xsl0rpr " + (size.xLargeUp != null ? "xpg4x96" : size.xLargeUp),
                ["$$css"]: true
            },
            {
                "--x-5qo2oh": size.default != null ? size.default : undefined,
                "--x-1ipfxcv": size.smallUp != null ? size.smallUp : undefined,
                "--x-dmxlzl": size.mediumUp != null ? size.mediumUp : undefined,
                "--x-1600g0r": size.largeUp != null ? size.largeUp : undefined,
                "--x-m6n0iu": size.xLargeUp != null ? size.xLargeUp : undefined
            }
        ]
};
