// @canva/app-ui-kit/dist/cjs/base/result.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "Result", {
    enumerable: true,
    get: function() {
        return Result;
    }
});
class OkImpl {
    get ok() {
        return true;
    }
    map(f) {
        return new OkImpl(f(this.value));
    }
    flatMap(f) {
        return f(this.value);
    }
    mapError() {
        return this;
    }
    get() {
        return this.value;
    }
    constructor(value){
        this.value = value;
    }
}
class ErrImpl {
    get ok() {
        return false;
    }
    map() {
        return this;
    }
    mapError(f) {
        return new ErrImpl(f(this.error));
    }
    flatMap() {
        return this;
    }
    get() {
        return undefined;
    }
    constructor(error){
        this.error = error;
    }
}
const Result = {
    Ok: OkFactory,
    Err: ErrFactory
};
const EMPTY_OK = new OkImpl(undefined);
const EMPTY_ERR = new ErrImpl(undefined);
function OkFactory(a) {
    return a === undefined ? EMPTY_OK : new OkImpl(a);
}
function ErrFactory(x) {
    return x === undefined ? EMPTY_ERR : new ErrImpl(x);
}
