// @canva/app-ui-kit/dist/esm/ui/apps/developing/ui_kit/components/typography/typography.js
import { jsx as _jsx } from "react/jsx-runtime";
import { UnreachableError } from '../../../../../../base/preconditions';
import 'react';
import { Text as EaselText, Title as EaselTitle } from '../../../../../base/typography/typography';
/** 
 * Renders body text.
 */ export function Text(props) {
    return _jsx(EaselText, {
        alignment: props.alignment,
        capitalization: props.capitalization,
        children: props.children,
        lineClamp: props.lineClamp,
        size: props.size,
        tagName: props.tagName,
        tone: mapTypographyTone(props.tone),
        weight: props.variant
    });
}
/** 
 * Renders heading text.
 */ export function Title(props) {
    return _jsx(EaselTitle, {
        alignment: props.alignment,
        capitalization: props.capitalization,
        children: props.children,
        lineClamp: props.lineClamp,
        size: props.size,
        tagName: props.tagName,
        tone: mapTypographyTone(props.tone)
    });
}
export function mapTypographyTone(tone) {
    switch(tone){
        case undefined:
            return undefined;
        case 'primary':
            return 'content';
        case 'secondary':
            return 'subtle';
        case 'tertiary':
            return 'subtlest';
        case 'critical':
            return 'critical';
        default:
            throw new UnreachableError(tone);
    }
}
