// @canva/app-ui-kit/dist/cjs/ui/apps/developing/ui_kit/components/surface_header/surface_header.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "SurfaceHeader", {
    enumerable: true,
    get: function() {
        return SurfaceHeader;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _header = require('../../../../../base/surface/header/header');
const _surface_headercss = _interop_require_default(require("./surface_header.css"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function SurfaceHeader(props) {
    const legacyPadding = '2u';
    return (0, _jsxruntime.jsx)("div", {
        className: _surface_headercss.default.legacy,
        children: (0, _jsxruntime.jsx)(_header.SurfaceHeader, {
            title: props.title,
            description: props.description,
            start: props.start && (0, _jsxruntime.jsx)(_header.SurfaceHeaderBackButton, {
                ariaLabel: props.start.ariaLabel,
                onClick: props.start.onClick
            }),
            divider: props.divider ?? true,
            background: props.background ?? 'surface',
            end: props.end,
            padding: legacyPadding
        })
    });
}
