// @canva/app-ui-kit/dist/esm/ui/base/box/private/box.js
import { jsx as _jsx } from "react/jsx-runtime";
import * as stylex from '../../../../platforms/libs/stylex/stylex';
import { toCallbackRef } from '../../../../base/react/callback_ref';
import classNames from 'classnames';
import 'react';
import { commonPropsToDOMProps } from '../common_props/common_props';
import checkeredPatternStyles from '../../checkered_pattern/checkered_pattern.css';
import { useBoxStyles } from './box_styles';
export function useBox({ className: classNameProp, customProperties, sx, ...options }) {
    const boxProps = stylex.props([
        useBoxStyles(options),
        sx
    ]);
    return {
        className: classNames(boxProps.className, classNameProp, getCheckeredPatternClassName(options)),
        style: {
            ...boxProps.style,
            ...customProperties
        },
        ['data-style-src']: boxProps['data-style-src']
    };
}
export function Box(props) {
    const { tagName: TagName = 'div', children, tabIndex, ref, ...rest } = props;
    return _jsx(TagName, {
        ref: toCallbackRef(ref),
        ...useBox(rest),
        ...commonPropsToDOMProps(rest),
        tabIndex: tabIndex,
        children: children
    });
}
function getCheckeredPatternClassName({ background }) {
    return background === 'checkered' ? checkeredPatternStyles.checkeredPattern : null;
}
