// @canva/app-ui-kit/dist/esm/ui/base/metrics/metrics.js
import { useExpectedPixelsPerRem } from '../provider/provider';
import styles from './private/metrics.css';
export { Breakpoint, breakpointFromScreenWidth, largeBreakpoint, mediumBreakpoint, smallBreakpoint, xLargeBreakpoint, aboveSmall, belowSmall, aboveMedium, belowMedium, aboveLarge, belowLarge, aboveXLarge, belowXLarge } from '../responsive/breakpoints';
export const baseUnit = parseInt(styles.baseUnit, 10) || 8;
export const unitSpaces = [
    '0',
    '0.25u',
    '0.5u',
    '0.75u',
    '1u',
    '1.25u',
    '1.5u',
    '2u',
    '3u',
    '4u',
    '6u',
    '8u',
    '12u'
];
export const getSpaceValue = (space)=>{
    return spaceValues[space];
};
const spaceValues = {
    ['0']: styles.space0,
    ['0.25u']: styles.space025,
    ['0.5u']: styles.space050,
    ['0.75u']: styles.space075,
    ['1u']: styles.space1,
    ['1.25u']: styles.space125,
    ['1.5u']: styles.space150,
    ['2u']: styles.space2,
    ['3u']: styles.space3,
    ['4u']: styles.space4,
    ['6u']: styles.space6,
    ['8u']: styles.space8,
    ['12u']: styles.space12
};
export const unitSizeToNumber = (unitSize)=>{
    return parseFloat(unitSize) * baseUnit;
};
export const resolveUnitOrNumberSize = (size)=>{
    if (size == null || typeof size === 'number') return size;
    return unitSizeToNumber(size);
};
export const DEFAULT_ROOT_FONT_SIZE = 10;
export const useBaseUnitInRem = ()=>{
    const rootFontSize = useExpectedPixelsPerRem() || DEFAULT_ROOT_FONT_SIZE;
    return baseUnit / rootFontSize;
};
export const getRootFontSize = ()=>{
    if (window == null) return DEFAULT_ROOT_FONT_SIZE;
    return parseFloat(getComputedStyle(window.document.documentElement).fontSize) || DEFAULT_ROOT_FONT_SIZE;
};
export const useTextZoomBaseUnit = (currentRootFontSize = getRootFontSize())=>{
    const originalRootFontSize = useExpectedPixelsPerRem() || DEFAULT_ROOT_FONT_SIZE;
    const textZoomLevel = currentRootFontSize / originalRootFontSize;
    return baseUnit * textZoomLevel;
};
