// eslint-plugin-formatjs/lib_esnext/rules/no-offset.js
import { isPluralElement, parse, } from '@formatjs/icu-messageformat-parser';
import { getParserServices } from '../context-compat';
import { extractMessages, getSettings } from '../util';
function verifyAst(ast) {
    const errors = [];
    for (const el of ast) {
        if (isPluralElement(el)) {
            if (el.offset) {
                errors.push({ messageId: 'noOffset', data: {} });
            }
            const { options } = el;
            for (const selector of Object.keys(options)) {
                errors.push(...verifyAst(options[selector].value));
            }
        }
    }
    return errors;
}
function checkNode(context, node) {
    const settings = getSettings(context);
    const msgs = extractMessages(node, settings);
    for (const [{ message: { defaultMessage }, messageNode, },] of msgs) {
        if (!defaultMessage || !messageNode) {
            continue;
        }
        const errors = verifyAst(parse(defaultMessage, {
            ignoreTag: settings.ignoreTag,
        }));
        for (const error of errors) {
            context.report({
                node: messageNode,
                ...error,
            });
        }
    }
}
export const name = 'no-offset';
export const rule = {
    meta: {
        type: 'problem',
        docs: {
            description: 'Disallow offset in plural rules',
            url: 'https://formatjs.github.io/docs/tooling/linter#no-offset',
        },
        fixable: 'code',
        messages: {
            noOffset: 'offset is not allowed',
        },
        schema: [],
    },
    defaultOptions: [],
    create(context) {
        const callExpressionVisitor = (node) => checkNode(context, node);
        const parserServices = getParserServices(context);
        //@ts-expect-error defineTemplateBodyVisitor exists in Vue parser
        if (parserServices?.defineTemplateBodyVisitor) {
            //@ts-expect-error
            return parserServices.defineTemplateBodyVisitor({
                CallExpression: callExpressionVisitor,
            }, {
                CallExpression: callExpressionVisitor,
            });
        }
        return {
            JSXOpeningElement: (node) => checkNode(context, node),
            CallExpression: callExpressionVisitor,
        };
    },
};
