// @canva/app-ui-kit/dist/esm/ui/base/form/input_decorators/private/button_decorator.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import classNames from 'classnames';
import * as React from 'react';
import { Badge } from '../../../badge/badge';
import { Button, CircleButton } from '../../../button/button';
import styles from './button_decorator.css';
export const ButtonDecorator = React.memo(
    function ButtonDecorator({ label, icon, badge, onClick, active, disclosure, disabled, className, ariaHasPopup, ariaControls, shape = 'regular' }) {
        const ButtonComponent = shape === 'round' ? CircleButton : Button;
        const button = _jsx(ButtonComponent, {
            size: shape === 'round' ? 'xsmall' : 'small',
            variant: "tertiary",
            iconSize: "medium",
            icon: icon,
            onClick: onClick,
            tooltipLabel: label,
            tooltipPlacement: "top",
            ariaLabel: label,
            ariaHasPopup: ariaHasPopup,
            ariaControls: ariaControls,
            active: active,
            disclosure: disclosure,
            disabled: disabled,
            className: classNames(badge != null && styles.enablePointerEvents, className)
        });
        if (badge != null) return _jsxs("div", {
            className: classNames(styles.badgeContainer, styles.disablePointerEvents),
            children: [
                button,
                _jsx("div", {
                    className: styles.badge,
                    children: typeof badge === 'string' ? _jsx(Badge, {
                        tone: "feedbackHint",
                        shape: "circle",
                        text: badge,
                        ariaLabel: badge
                    }) : badge
                })
            ]
        });
        return button;
    }
);
