// @canva/app-ui-kit/dist/cjs/ui/base/thumbnail/private/thumbnail_container.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get SizingContainer () {
        return SizingContainer;
    },
    get ThumbnailContainer () {
        return ThumbnailContainer;
    },
    get ThumbnailContainerContext () {
        return ThumbnailContainerContext;
    },
    get resolveAspectRatio () {
        return resolveAspectRatio;
    },
    get useInThumbnailContainerContext () {
        return useInThumbnailContainerContext;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _legacy_class_name = require('../../../../platforms/libs/stylex/legacy_class_name');
const _react = _interop_require_wildcard(require("react"));
const _aspect_ratio = require('../../aspect_ratio/aspect_ratio');
const _metrics = require('../../metrics/metrics');
require('../../tokens/primitive_tokens.stylex');
require('../../tokens/semantic_tokens.stylex');
const _thumbnail_styles = require("./thumbnail_styles");
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const ThumbnailContainerContext = _react.createContext(undefined);
const useInThumbnailContainerContext = ()=>{
    const containerContext = _react.useContext(ThumbnailContainerContext);
    return containerContext != null;
};
const ThumbnailContainer = ({ children, className, background = 'secondary', border = 'none', borderRadius = 'elementSmall', fit = 'cover', padding = '2u', height, width, aspectRatio: aspectRatioProp })=>{
    const aspectRatio = resolveAspectRatio({
        width,
        height,
        aspectRatio: aspectRatioProp
    });
    return (0, _jsxruntime.jsx)(SizingContainer, {
        height: height,
        width: width,
        aspectRatio: aspectRatio,
        children: (0, _jsxruntime.jsx)("div", {
            ...(0, _legacy_class_name.composeLegacyClassNamePropWithStyles)(className, [
                styles.root,
                ...getThumbnailContainerStyles({
                    background,
                    border,
                    borderRadius,
                    padding
                })
            ]),
            children: (0, _jsxruntime.jsx)(ThumbnailContainerContext, {
                value: {
                    aspectRatio,
                    padding,
                    fit
                },
                children: children
            })
        })
    });
};
const getThumbnailContainerStyles = ({ background, border, borderRadius, padding })=>[
        background === 'secondary' && styles.secondaryBackground,
        background === 'contrastOnLight' && styles.contrastOnLight,
        background === 'contrastOnDark' && styles.contrastOnDark,
        background === 'secondaryHover' && styles.secondaryBackgroundHover,
        background === 'contrastOnLightHover' && styles.contrastOnLightHover,
        background === 'contrastOnDarkHover' && styles.contrastOnDarkHover,
        border !== 'none' && _thumbnail_styles.sharedThumbnailStyles.border,
        border === 'low' && _thumbnail_styles.sharedThumbnailStyles.lowBorder,
        borderRadius === 'elementSmall' && _thumbnail_styles.sharedThumbnailStyles.elementSmallBorderRadius,
        borderRadius === 'element' && _thumbnail_styles.sharedThumbnailStyles.elementBorderRadius,
        borderRadius === 'elementRelaxed' && _thumbnail_styles.sharedThumbnailStyles.elementRelaxedBorderRadius,
        borderRadius === 'elementSoftest' && _thumbnail_styles.sharedThumbnailStyles.elementSoftestBorderRadius,
        borderRadius === 'elementRound' && _thumbnail_styles.sharedThumbnailStyles.elementRoundBorderRadius,
        padding === '1u' && styles.paddingOneUnit,
        padding === '2u' && styles.paddingTwoUnit
    ];
const SizingContainer = ({ aspectRatio, width, height, children })=>{
    const widthInPixel = (0, _metrics.resolveUnitOrNumberSize)(width);
    const heightInPixel = (0, _metrics.resolveUnitOrNumberSize)(height);
    if (aspectRatio != null) return (0, _jsxruntime.jsx)(_aspect_ratio.AspectRatio, {
        ratio: aspectRatio,
        width: width,
        height: height,
        children: children
    });
    if (width != null || height != null) return (0, _jsxruntime.jsx)("div", {
        style: {
            width: widthInPixel ?? '100%',
            height: heightInPixel ?? '100%',
            flexShrink: 0
        },
        children: children
    });
    return children;
};
const resolveAspectRatio = ({ aspectRatio, width, height })=>{
    const widthInPixel = (0, _metrics.resolveUnitOrNumberSize)(width);
    const heightInPixel = (0, _metrics.resolveUnitOrNumberSize)(height);
    return aspectRatio ?? (widthInPixel != null && heightInPixel != null ? widthInPixel / heightInPixel : undefined);
};
const styles = {
    root: {
        kVQacm: "xb3r6kr",
        kGNEyG: "x6s0dn4",
        k1xSpc: "x78zum5",
        kjj79g: "xl56j7k",
        kZKoxP: "x5yr21d",
        kzqmXN: "xh8yej3",
        ["$$css"]: true
    },
    secondaryBackground: {
        kWkggS: "x1gim84k",
        ["$$css"]: true
    },
    secondaryBackgroundHover: {
        kWkggS: "xbdlxvi",
        ["$$css"]: true
    },
    contrastOnLight: {
        kWkggS: "x110c1mw",
        ["$$css"]: true
    },
    contrastOnLightHover: {
        kWkggS: "xqxvuf3",
        ["$$css"]: true
    },
    contrastOnDark: {
        kWkggS: "x9eh2el",
        ["$$css"]: true
    },
    contrastOnDarkHover: {
        kWkggS: "xong8os",
        ["$$css"]: true
    },
    paddingOneUnit: {
        kmVPX3: "xmom2u2",
        kB7OPa: "x9f619",
        ["$$css"]: true
    },
    paddingTwoUnit: {
        kmVPX3: "x5vm2tf",
        kB7OPa: "x9f619",
        ["$$css"]: true
    }
};
