// @canva/app-ui-kit/dist/esm/ui/base/form/select/private/stateless_base_select.js
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import { composeRefs } from '../../../../../base/react/compose_refs';
import classNames from 'classnames';
import * as React from 'react';
import { useIsKeyboardAccessMode } from '../../../a11y/access_mode_state/access_mode_state';
import { useAnnouncer } from '../../../a11y/announcer/announcer';
import { ScreenReaderContent } from '../../../a11y/screen_reader_content/screen_reader_content';
import { Box } from '../../../box/box';
import { BaseButton } from '../../../button/base_button/base_button';
import { Divider } from '../../../divider/divider';
import { FakeCheckbox } from '../../checkbox/checkbox';
import { composeHandles } from '../../../handle/handle';
import { CheckIcon } from '../../../icons/check/icon';
import { ChevronDownIcon } from '../../../icons/chevron_down/icon';
import { Rows, Spacer } from '../../../layout/layout';
import { DIVIDER_HEIGHT_MULTIPIER, ITEM_HEIGHT_MULTIPIER, ITEM_WITH_DESCRIPTION_HEIGHT_MULTIPIER, Menu as BaseMenu, MenuDivider, MenuItem, TEXT_DIVIDER_HEIGHT_MULTIPIER } from '../../../menu/menu';
import { getRootFontSize, useTextZoomBaseUnit } from '../../../metrics/metrics';
import { useEnableHighColorContrast } from '../../../provider/provider';
import { isAboveView, isBelowView, scrollIntoView } from '../../../scroll/scroll';
import { useScrollable } from '../../../scroll/scrollable/scrollable';
import { useVirtualizedList } from '../../../scroll/virtualized_list/virtualized_list';
import { SurfaceHeader } from '../../../surface/header/header';
import { useTypicalSheetAdaptation } from '../../../surface/sheet/sheet';
import colorTokens from '../../../tokens/color.css';
import { InheritColor, Text } from '../../../typography/typography';
import styles, { getStyle } from './base_select.css';
import { BaseSelectMessages } from './base_select.messages';
import { BaseSelectTestIds } from './base_select.testids';
import { defaultAreEqual, defaultGetLabel, defaultIsSelected, deriveIdentifier, flattenOptions, joinLabels, normalizeOptionGroups, sliceOptionGroups } from './base_select_util';
import { useDefaultSelectRenderSurface } from './surface';
import { TruncatedList } from './truncated_list';
export function StatelessBaseSelect(props) {
    const { Trigger = BaseSelectTrigger, Menu = BaseSelectMenu, Item, renderSurface, trigger = (props)=> _jsx(Trigger, {
            ...props
        }), menu = (props)=> _jsx(Menu, {
            ...props
        }), item, value, placeholder, open, disabled, error, stretch, activeIndex, options, size, focusOnMenu = false, isMenuContainerFocusable = false, id, ariaLabelledBy, ariaDescribedBy, tooltipLabel, tooltipPlacement, tooltipDisabled, className, onRequestClose, onRequestOpen, onBlur, onFocus, onItemClick, onItemHover, areEqual = defaultAreEqual, isSelected = defaultIsSelected, getLabel = defaultGetLabel } = props;
    const menuId = React.useId();
    const labelId = React.useId();
    const sizingMode = useTypicalSheetAdaptation() ? 'fill' : 'fixed';
    const isKeyboardMode = useIsKeyboardAccessMode();
    const triggerContainerRef = React.useRef(null);
    const menuContainerRef = React.useRef(null);
    const sheetScrollRef = React.useRef(null);
    const sheetHandle = React.useCallback((handle)=>{
        if (handle != null) sheetScrollRef.current = sheetHandleToScrollContainer(handle);
        return ()=>{
            sheetScrollRef.current = null;
        };
    }, [
        sheetScrollRef
    ]);
    const flatOptions = React.useMemo(()=>props.flatOptions ?? flattenOptions(props.options), [
        props.options,
        props.flatOptions
    ]);
    const defaultRenderSurface = useDefaultSelectRenderSurface({
        optionCount: flatOptions.length
    });
    const optionGroups = React.useMemo(()=>props.optionGroups ?? normalizeOptionGroups(props.options), [
        props.options,
        props.optionGroups
    ]);
    const selectedOptions = React.useMemo(()=>props.selectedOptions ?? flatOptions.filter((option)=>isSelected(option, value)), [
        flatOptions,
        props.selectedOptions,
        isSelected,
        value
    ]);
    const activeDescendantId = open ? deriveIdentifier(menuId, activeIndex) : undefined;
    const ariaLabel = props.ariaLabel ?? (!ariaLabelledBy && typeof props.title === 'string' ? props.title : undefined);
    const onRequestToggle = React.useCallback(()=>{
        if (open) onRequestClose();
        else onRequestOpen();
    }, [
        open,
        onRequestClose,
        onRequestOpen
    ]);
    const header = props.title ? _jsx(SurfaceHeader, {
        title: props.title,
        end: props.headerEnd
    }) : undefined;
    const [surfaceScrolled, setSurfaceScrolled] = React.useState(false);
    const handleSurfaceScroll = React.useCallback(({ scrollTop })=>{
        setSurfaceScrolled(scrollTop > 0);
    }, []);
    React.useEffect(()=>{
        if (!open) setSurfaceScrolled(false);
    }, [
        open
    ]);
    const triggerContent = _jsx("div", {
        ref: composeRefs(triggerContainerRef, props.triggerContainerRef),
        className: classNames(styles.triggerContainer, className, {
            [styles.triggerStretch]: stretch,
            [styles.triggerOutline]: focusOnMenu && open && isKeyboardMode
        }),
        "data-testid": BaseSelectTestIds.triggerContainer,
        children: trigger({
            value,
            options,
            flatOptions,
            optionGroups,
            selectedOptions,
            placeholder,
            disabled,
            error,
            open,
            id,
            ariaLabel,
            ariaLabelledBy,
            ariaDescribedBy,
            activeIndex,
            onBlur,
            onFocus,
            role: 'combobox',
            disclosure: true,
            labelId,
            active: open,
            ariaActiveDescendant: activeDescendantId,
            ariaControls: open ? menuId : undefined,
            onRequestToggle,
            getLabel,
            isSelected,
            size,
            tooltipLabel,
            tooltipPlacement,
            tooltipDisabled
        })
    });
    const menuContent = _jsx("div", {
        className: styles.menuContainer,
        ref: composeRefs(menuContainerRef, props.menuContainerRef),
        "data-testid": BaseSelectTestIds.menuContainer,
        tabIndex: isMenuContainerFocusable ? 0 : undefined,
        "aria-activedescendant": isMenuContainerFocusable ? activeDescendantId : undefined,
        children: menu({
            Item,
            item,
            id: menuId,
            options,
            flatOptions,
            optionGroups,
            selectedOptions,
            value,
            activeIndex,
            activeOption: flatOptions[activeIndex ?? -1],
            onItemClick,
            onItemHover,
            areEqual,
            isSelected,
            getLabel,
            sizingMode,
            menuContainerRef,
            ariaLabel,
            ariaLabelledBy,
            isScrolled: surfaceScrolled,
            sheetScrollRef,
            children: options.length === 0 ? renderEmptyMenuContent() : undefined
        })
    });
    const surfaceProps = {
        open: open ?? false,
        onRequestClose,
        captureFocus: focusOnMenu,
        header,
        headerDivider: 'always',
        onScroll: handleSurfaceScroll,
        trigger: triggerContent,
        children: menuContent,
        sheetHandle
    };
    return _jsx("div", {
        children: (renderSurface ?? defaultRenderSurface)(surfaceProps)
    });
}
export function BaseSelectTrigger(props) {
    const { selectedOptions } = props;
    const noSelection = !selectedOptions.length;
    const Icon = getIcon(props.Icon ?? 'auto', selectedOptions[0]?.Icon);
    const size = props.size ?? 'medium';
    const ariaLabelledBy = props.ariaLabelledBy || (props.ariaLabel ? undefined : props.labelId);
    const ariaDescribedBy = getTriggerAriaDescribedBy(props);
    return _jsxs(BaseButton, {
        onClick: props.onRequestToggle,
        onBlur: props.onBlur,
        onFocus: props.onFocus,
        disclosure: true,
        pressed: props.open,
        disabled: props.disabled,
        className: classNames(styles.triggerButton, getStyle(size), props.className),
        id: props.id,
        role: props.role,
        tooltipLabel: props.tooltipLabel,
        tooltipPlacement: props.tooltipPlacement,
        tooltipDisabled: props.tooltipDisabled,
        ariaLabel: props.ariaLabel,
        ariaLabelledBy: ariaLabelledBy,
        ariaDescribedBy: ariaDescribedBy,
        ariaControls: props.ariaControls,
        ariaHasPopup: "listbox",
        ariaActiveDescendant: props.ariaActiveDescendant,
        ariaInvalid: props.error,
        borderRadius: "elementStandard",
        paddingX: size === 'medium' ? '1.5u' : '2u',
        alignItems: "center",
        width: "full",
        border: "control",
        tone: "secondary",
        backgroundColor: {
            default: colorTokens.colorControlBg,
            pressed: colorTokens.colorControlBg,
            hovered: colorTokens.colorControlBg,
            disabled: colorTokens.colorControlBgDisabled
        },
        borderColor: {
            default: props.error ? colorTokens.colorControlCriticalBorder : colorTokens.colorControlBorder,
            pressed: colorTokens.colorControlBorderFocused,
            hovered: colorTokens.colorControlBorderHovered,
            disabled: colorTokens.colorControlBorderDisabled
        },
        children: [
            Icon && _jsxs(_Fragment, {
                children: [
                    _jsx(Icon, {
                        size: "medium"
                    }),
                    _jsx(Spacer, {
                        size: "1u"
                    })
                ]
            }),
            _jsx(Text, {
                id: props.labelId,
                tagName: "span",
                lineClamp: 1,
                tone: InheritColor,
                className: classNames(styles.label, {
                    [styles.placeholder]: noSelection || props.disabled
                }),
                children: getTriggerContent(props)
            }),
            _jsx(ChevronDownIcon, {
                size: "medium"
            })
        ]
    });
}
function getIcon(Icon, FallbackIcon) {
    if (Icon === 'none') return undefined;
    if (Icon === 'auto') return FallbackIcon;
    return Icon;
}
function getTriggerContent(props) {
    if (props.children) return props.children;
    const hasSelection = props.selectedOptions.length > 0;
    if (!hasSelection && props.disabled)
    return _jsxs(_Fragment, {
        children: [
            _jsx(ScreenReaderContent, {
                tagName: "span",
                children: BaseSelectMessages.defaultDisabledMessage()
            }),
            _jsx("span", {
                "aria-hidden": true,
                children: props.placeholder ?? BaseSelectMessages.defaultPlaceholder()
            })
        ]
    });
    if (!hasSelection)
    return _jsxs(_Fragment, {
        children: [
            _jsx(ScreenReaderContent, {
                tagName: "span",
                children: BaseSelectMessages.defaultNoSelectionMessage()
            }),
            _jsx("span", {
                "aria-hidden": true,
                id: `${props.labelId}--placeholder`,
                children: props.placeholder ?? BaseSelectMessages.defaultPlaceholder()
            })
        ]
    });
    const items = props.selectedOptions.map(props.getLabel);
    if (items.length > 1 && items.every((item)=>typeof item === 'string')) return _jsx(TruncatedList, {
        tagName: "span",
        items: items
    });
    return joinLabels(items);
}
function getTriggerAriaDescribedBy(props) {
    if (props.children)
    return props.ariaDescribedBy;
    const hasSelection = props.selectedOptions.length > 0;
    if (!hasSelection) return `${props.ariaDescribedBy ?? ''} ${props.labelId}--placeholder`.trim();
    return props.ariaDescribedBy;
}
const sheetHandleToScrollContainer = (sheetHandle)=>{
    if (sheetHandle == null) return null;
    return composeHandles(sheetHandle, {
        scrollTo: (optionsOrX, y)=>{
            if (typeof optionsOrX === 'number') sheetHandle.scrollTo({
                left: optionsOrX,
                top: y
            });
            else if (optionsOrX != null) sheetHandle.scrollTo(optionsOrX);
        },
        scrollBy: (optionsOrX, y)=>{
            if (typeof optionsOrX === 'number') sheetHandle.scrollBy({
                left: optionsOrX,
                top: y
            });
            else if (optionsOrX != null) sheetHandle.scrollBy(optionsOrX);
        },
        scrollIntoView: ()=>{}
    });
};
function useScrollToActiveItem({ activeIndex, scrollRef: scrollRefProp }) {
    const scrollRef = React.useRef(null);
    const activeItemRef = React.useRef(null);
    const scrollContainer = scrollRefProp ?? scrollRef;
    React.useEffect(()=>{
        if (!scrollContainer.current || !activeItemRef.current) return;
        if (isAboveView(scrollContainer.current, activeItemRef.current)) scrollIntoView(scrollContainer.current, activeItemRef.current, {
            align: 'start'
        });
        if (isBelowView(scrollContainer.current, activeItemRef.current)) scrollIntoView(scrollContainer.current, activeItemRef.current, {
            align: 'end'
        });
    }, [
        activeIndex
    ]);
    return {
        activeItemRef,
        scrollRef
    };
}
function useSelectScrollable(onScrollStateChange) {
    const [isScrolled, setIsScrolled] = React.useState(false);
    const WrappedOnScrollStateChange = React.useCallback((scrollState)=>{
        onScrollStateChange?.(scrollState);
        setIsScrolled(!scrollState.atTop);
    }, [
        onScrollStateChange
    ]);
    const { ref, onScroll } = useScrollable({
        onScrollStateChange: WrappedOnScrollStateChange
    });
    return {
        isScrolled,
        ref,
        onScroll
    };
}
export function BaseSelectMenu(props) {
    const { Item = BaseSelectItem, item = (props)=> _jsx(Item, {
            ...props
        }), flatOptions, optionGroups, value, activeIndex, id, ariaLabel, ariaLabelledBy, header, footer, isSelected, getLabel, onItemClick, onItemHover, className, children, sizingMode, isScrolled: isScrolledProp, sheetScrollRef } = props;
    const isPopoverMode = sizingMode === 'fixed';
    const { isScrolled: isPopoverScrolled, ref: popoverScrollRef, onScroll: popoverOnScroll } = useSelectScrollable();
    const { activeItemRef, scrollRef: scrollToActiveItemRef } = useScrollToActiveItem({
        scrollRef: isPopoverMode ? undefined : sheetScrollRef,
        activeIndex
    });
    const isScrolled = isPopoverMode ? isPopoverScrolled : isScrolledProp;
    const hasMultipleGroups = optionGroups.length > 1;
    const announcer = useAnnouncer();
    const announcerRef = React.useRef(null);
    const announcedContent = React.useRef('');
    const shouldAnnounce = !!children;
    React.useEffect(()=>{
        if (!shouldAnnounce) return;
        const textContent = announcerRef.current?.textContent ?? '';
        if (textContent === announcedContent.current) return;
        announcedContent.current = textContent;
        announcer.announce(textContent);
    }, [
        announcer,
        shouldAnnounce,
        children
    ]);
    return _jsx(WithHeaderAndFooter, {
        header: header,
        footer: footer,
        sizingMode: sizingMode,
        isScrolled: isScrolled,
        children: _jsx("div", {
            className: classNames(styles.scrollContainer, sizingMode === 'fixed' && styles.popover, sizingMode === 'fill' && !header && styles.sheetNoHeader, className),
            ref: isPopoverMode ? composeRefs(scrollToActiveItemRef, popoverScrollRef) : undefined,
            onScroll: isPopoverMode ? popoverOnScroll : undefined,
            id: hasMultipleGroups ? id : undefined,
            role: hasMultipleGroups ? 'listbox' : undefined,
            "aria-label": hasMultipleGroups ? ariaLabel : undefined,
            "aria-labelledby": hasMultipleGroups ? ariaLabelledBy : undefined,
            children: children ? _jsx("div", {
                ref: announcerRef,
                children: children
            }) : optionGroups.map((optionGroup, groupIndex)=>_jsxs(React.Fragment, {
                    children: [
                        groupIndex > 0 && _jsx("div", {
                            className: styles.groupDivider,
                            children: _jsx(Divider, {})
                        }),
                        _jsx(OptionGroup, {
                            item: item,
                            menuId: id,
                            id: hasMultipleGroups ? undefined : id,
                            role: hasMultipleGroups ? 'group' : 'listbox',
                            ariaLabel: hasMultipleGroups ? undefined : ariaLabel,
                            ariaLabelledBy: hasMultipleGroups ? undefined : ariaLabelledBy,
                            activeItemRef: activeItemRef,
                            value: value,
                            optionIndexOffset: flatOptions.findIndex((o)=>o === optionGroup.options[0]),
                            activeIndex: activeIndex,
                            label: optionGroup.label,
                            options: optionGroup.options,
                            isSelected: isSelected,
                            onItemClick: onItemClick,
                            onItemHover: onItemHover,
                            getLabel: getLabel
                        })
                    ]
                }, groupIndex))
        })
    });
}
function OptionGroup({ item = (props)=> _jsx(BaseSelectItem, {
        ...props
    }), id, menuId, role, ariaLabel, ariaLabelledBy, value, optionIndexOffset = 0, activeIndex, label, options, isSelected, activeItemRef, onItemClick, onItemHover, getLabel }) {
    const headingId = React.useId();
    const multiSelectable = Array.isArray(value);
    return _jsxs(BaseMenu, {
        id: id,
        role: role,
        ariaMultiSelectable: multiSelectable,
        ariaLabel: ariaLabel,
        ariaLabelledBy: label ? headingId : ariaLabelledBy,
        children: [
            label != null && _jsx(MenuDivider, {
                id: headingId,
                children: label
            }),
            options.map((option, i)=>{
                const selected = isSelected(option, value);
                const itemIndex = optionIndexOffset + i;
                const active = activeIndex === itemIndex;
                const id = deriveIdentifier(menuId, itemIndex);
                return _jsx(React.Fragment, {
                    children: item({
                        id,
                        selected,
                        active,
                        forwardedRef: active ? activeItemRef : undefined,
                        onItemClick,
                        onItemHover,
                        getLabel,
                        option,
                        multiSelectable
                    })
                }, id);
            })
        ]
    });
}
export function WindowedSelectMenu(props) {
    const { Item = BaseSelectItem, item = (props)=> _jsx(Item, {
            ...props
        }), optionGroups, flatOptions, value, activeIndex, id, ariaLabel, ariaLabelledBy, header, footer, isSelected, getLabel, onItemClick, onItemHover, className, children, sizingMode, isScrolled: isScrolledProp, itemHeightMultiplier: itemHeightMultiplierProp, sheetScrollRef } = props;
    const rootFontSize = React.useMemo(()=>{
        return getRootFontSize();
    }, []);
    const textZoomBaseUnit = useTextZoomBaseUnit(rootFontSize);
    const itemSizes = React.useMemo(()=>{
        const itemSizes = [];
        optionGroups.forEach((optionGroup, groupIndex)=>{
            optionGroup.options.forEach((option, optionIndex)=>{
                const absoluteIndex = itemSizes.length;
                const providedItemHeightMultiplier = typeof itemHeightMultiplierProp === 'function' ? itemHeightMultiplierProp(option, absoluteIndex) : itemHeightMultiplierProp;
                const itemHeightMultiplier = providedItemHeightMultiplier ?? (option.description ? ITEM_WITH_DESCRIPTION_HEIGHT_MULTIPIER : ITEM_HEIGHT_MULTIPIER);
                let height = itemHeightMultiplier * textZoomBaseUnit;
                if (optionIndex === 0 && groupIndex > 0)
                    height += DIVIDER_HEIGHT_MULTIPIER * textZoomBaseUnit;
                if (optionIndex === 0 && optionGroup.label != null)
                    height += TEXT_DIVIDER_HEIGHT_MULTIPIER * textZoomBaseUnit;
                itemSizes.push(height);
            });
        });
        return itemSizes;
    }, [
        itemHeightMultiplierProp,
        optionGroups,
        textZoomBaseUnit
    ]);
    const isPopover = sizingMode === 'fixed';
    const getItemSize = React.useCallback((index)=>{
        return itemSizes[index];
    }, [
        itemSizes
    ]);
    const { firstRenderedIndex, lastRenderedIndex, sizeBefore, sizeAfter, onScrollStateChange } = useVirtualizedList({
        getItemSize,
        gap: 0,
        itemCount: itemSizes.length
    });
    const visibleGroups = sliceOptionGroups(optionGroups, firstRenderedIndex, lastRenderedIndex + 1);
    const { isScrolled: isPopoverScrolled, ref: scrollableRef, onScroll } = useSelectScrollable(onScrollStateChange);
    const isScrolled = isPopover ? isPopoverScrolled : isScrolledProp;
    const { activeItemRef, scrollRef: scrollToActiveItemRef } = useScrollToActiveItem({
        activeIndex,
        scrollRef: isPopover ? undefined : sheetScrollRef
    });
    const scrollContainer = isPopover ? scrollToActiveItemRef : sheetScrollRef;
    React.useEffect(()=>{
        if (!scrollContainer || !scrollContainer.current || activeIndex == null || activeIndex < 5)
            return;
        const activeOptionHeight = itemSizes[activeIndex];
        const activeOptionPosition = itemSizes.slice(0, activeIndex).reduce((a, h)=>a + h);
        let paddingTop = 10;
        if (isPopover)
            paddingTop = parseInt(window.getComputedStyle(scrollToActiveItemRef.current).paddingTop, 10);
        const scrollOffset = scrollContainer.current.offsetHeight - paddingTop - activeOptionHeight;
        scrollContainer.current.scrollTo(0, activeOptionPosition - scrollOffset);
    }, []);
    const hasMultipleGroups = optionGroups.length > 1;
    return _jsx(WithHeaderAndFooter, {
        header: header,
        footer: footer,
        sizingMode: sizingMode,
        isScrolled: isScrolled,
        children: _jsx("div", {
            className: classNames(styles.scrollContainer, sizingMode === 'fixed' && styles.popover, sizingMode === 'fill' && !header && styles.sheetNoHeader, className),
            ref: isPopover ? composeRefs(scrollableRef, scrollToActiveItemRef) : scrollableRef,
            onScroll: onScroll,
            id: hasMultipleGroups ? id : undefined,
            role: hasMultipleGroups ? 'listbox' : undefined,
            "aria-label": hasMultipleGroups ? ariaLabel : undefined,
            "aria-labelledby": hasMultipleGroups ? ariaLabelledBy : undefined,
            children: children ? _jsx("div", {
                role: "alert",
                children: children
            }) : _jsxs(_Fragment, {
                children: [
                    _jsx("div", {
                        role: "none",
                        style: {
                            height: sizeBefore
                        }
                    }),
                    visibleGroups.map((optionGroup)=>{
                        const isFirstItemVisible = optionGroup.slice[0] === 0;
                        return _jsxs(React.Fragment, {
                            children: [
                                optionGroup.originalIndex > 0 && isFirstItemVisible && _jsx("div", {
                                    className: styles.groupDivider,
                                    children: _jsx(Divider, {})
                                }),
                                _jsx(OptionGroup, {
                                    item: item,
                                    menuId: id,
                                    id: hasMultipleGroups ? undefined : id,
                                    role: hasMultipleGroups ? 'group' : 'listbox',
                                    ariaLabel: hasMultipleGroups ? optionGroup.label : ariaLabel,
                                    ariaLabelledBy: hasMultipleGroups ? undefined : ariaLabelledBy,
                                    activeItemRef: activeItemRef,
                                    value: value,
                                    optionIndexOffset: flatOptions.findIndex((o)=>o === optionGroup.options[0]),
                                    activeIndex: activeIndex,
                                    label: isFirstItemVisible ? optionGroup.label : undefined,
                                    options: optionGroup.options,
                                    isSelected: isSelected,
                                    onItemClick: onItemClick,
                                    onItemHover: onItemHover,
                                    getLabel: getLabel
                                })
                            ]
                        }, optionGroup.originalIndex);
                    }),
                    _jsx("div", {
                        role: "none",
                        style: {
                            height: sizeAfter
                        }
                    })
                ]
            })
        })
    });
}
function WithHeaderAndFooter(props) {
    const { header, footer, children, sizingMode = 'fixed', isScrolled = false } = props;
    const isRenderFunctionHeader = typeof header === 'function';
    const headerContent = isRenderFunctionHeader ? header({
        isScrolled,
        sizingMode
    }) : typeof header === 'string' ? _jsx(Text, {
        children: header
    }) : header;
    return _jsxs(_Fragment, {
        children: [
            headerContent && _jsx("header", {
                className: classNames(styles.menuHeader, !isRenderFunctionHeader && styles.menuHeaderDefault),
                children: headerContent
            }),
            children,
            footer && _jsx("footer", {
                className: styles.menuFooter,
                children: typeof footer === 'string' ? _jsx(Text, {
                    children: footer
                }) : footer
            })
        ]
    });
}
export function renderEmptyMenuContent(searchQuery) {
    if (searchQuery) return _jsx(Box, {
        padding: "3u",
        children: _jsxs(Rows, {
            spacing: "0.5u",
            children: [
                _jsx(Text, {
                    alignment: "center",
                    weight: "bold",
                    children: BaseSelectMessages.noResultsFor(searchQuery)
                }),
                _jsx(Text, {
                    alignment: "center",
                    tone: "subtle",
                    children: BaseSelectMessages.tryAnotherTerm()
                })
            ]
        })
    });
    return _jsx(Box, {
        padding: "3u",
        children: _jsx(Text, {
            children: BaseSelectMessages.defaultEmptyMessage()
        })
    });
}
export const BaseSelectItem = React.memo(BaseSelectItem_);
function BaseSelectItem_(props) {
    const { onItemClick, onItemHover, option, multiSelectable } = props;
    const highContrast = useEnableHighColorContrast();
    const onClick = React.useCallback((event)=>{
        onItemClick?.(option, event);
    }, [
        onItemClick,
        option
    ]);
    const onMouseEnter = React.useCallback((event)=>{
        onItemHover?.(option, event);
    }, [
        onItemHover,
        option
    ]);
    const OptionIcon = option.Icon && _jsx(option.Icon, {
        size: "medium"
    });
    const SelectedIcon = !multiSelectable ? _jsxs("span", {
        "aria-hidden": true,
        children: [
            _jsx(Spacer, {
                direction: "horizontal",
                size: "1u"
            }),
            props.selected ? _jsx(CheckIcon, {
                size: "medium"
            }) : _jsx(Spacer, {
                direction: "horizontal",
                size: "3u"
            })
        ]
    }) : undefined;
    const CheckboxIcon = _jsx(FakeCheckbox, {
        checked: props.selected,
        disabled: props.option.disabled
    });
    const start = multiSelectable ? _jsxs(_Fragment, {
        children: [
            CheckboxIcon,
            props.start ?? OptionIcon
        ]
    }) : props.start ?? OptionIcon;
    const end = props.end ?? SelectedIcon;
    const content = props.children ?? (props.option.description != null ? _jsxs(Rows, {
        tagName: "span",
        spacing: "0",
        children: [
            _jsx(Text, {
                tagName: "span",
                lineClamp: 1,
                tone: InheritColor,
                children: props.getLabel(props.option)
            }),
            _jsx(Text, {
                tagName: "span",
                size: "small",
                tone: option.disabled ? InheritColor : 'subtlest',
                children: props.option.description
            })
        ]
    }) : props.getLabel(props.option));
    return _jsx(MenuItem, {
        id: props.id,
        ariaSelected: props.selected,
        buttonRef: props.forwardedRef,
        tabIndex: -1,
        onClick: onClick,
        onMouseEnter: onMouseEnter,
        start: start,
        end: end,
        className: classNames(props.className, {
            [styles.active]: props.active,
            [styles.disabled]: props.option.disabled,
            [styles.highContrast]: highContrast
        }),
        disabled: props.option.disabled,
        children: content
    });
}
