// @canva/app-ui-kit/dist/esm/ui/base/layout/private/spacer/spacer.js
import { jsx as _jsx } from "react/jsx-runtime";
import * as stylex from '../../../../../platforms/libs/stylex/stylex';
import 'react';
import { getSpaceValue } from '../../../metrics/metrics';
import { mapResponsiveValueToStyleX } from '../../../responsive/responsive';
import { SpacerTestIds } from './spacer.testids';
import './vars.stylex';
export const Spacer = ({ size, direction = 'both' })=>_jsx("span", {
        ...stylex.props([
            styles.root,
            direction === 'horizontal' && styles.horizontal,
            direction === 'vertical' && styles.vertical,
            mapResponsiveValueToStyleX(size, getSpaceValue, styles.size, styles.sizeResponsive)
        ]),
        "aria-hidden": true,
        "data-testid": SpacerTestIds.spacer
    });
const styles = {
    root: {
        k1xSpc: "x1lliihq",
        kmuXW: "x2lah0s",
        kZKoxP: "x11lv39n",
        kzqmXN: "x1ca76lx",
        ["$$css"]: true
    },
    horizontal: {
        k1xSpc: "x1rg5ohu",
        kZKoxP: "xt7dq6l",
        kAzted: "x4zw9a1",
        ["$$css"]: true
    },
    vertical: {
        k7Eaqz: "xgc0pbm",
        kzqmXN: "x14atkfc",
        ["$$css"]: true
    },
    size: (size)=>[
            {
                "--xqvgnxq": size != null ? "x1lahtpn" : size,
                ["$$css"]: true
            },
            {
                "--x---xqvgnxq": size != null ? size : undefined
            }
        ],
    sizeResponsive: (size)=>[
            {
                "--xqvgnxq": (size.default != null ? "xtlhp39 " : size.default) + "x4h0bxt " + "x1f3rlcv " + "xsl0rpr " + (size.xLargeUp != null ? "xpg4x96" : size.xLargeUp),
                ["$$css"]: true
            },
            {
                "--x-5qo2oh": size.default != null ? size.default : undefined,
                "--x-1ipfxcv": size.smallUp != null ? size.smallUp : undefined,
                "--x-dmxlzl": size.mediumUp != null ? size.mediumUp : undefined,
                "--x-1600g0r": size.largeUp != null ? size.largeUp : undefined,
                "--x-m6n0iu": size.xLargeUp != null ? size.xLargeUp : undefined
            }
        ]
};
