// @canva/app-ui-kit/dist/esm/ui/apps/developing/ui_kit/components/surface_header/surface_header.js
import { jsx as _jsx } from "react/jsx-runtime";
import { SurfaceHeader as EaselSurfaceHeader, SurfaceHeaderBackButton as EaselSurfaceHeaderBackButton } from '../../../../../base/surface/header/header';
import styles from './surface_header.css';
/** 
 * Display a header with a title, description, and start and end decorators such as back and close buttons.
 * @example
 * ```tsx
 * <SurfaceHeader
 *   title="Title"
 *   description="Description"
 *   divider={false}
 *   start={{ ariaLabel: 'Go back', onClick: () => {...} },
 *   end={ <Button variant="tertiary" icon={() => <MoreHorizontalIcon />} /> }
 * />
 * ```
 */ export function SurfaceHeader(props) {
    const legacyPadding = '2u';
    return _jsx("div", {
        className: styles.legacy,
        children: _jsx(EaselSurfaceHeader, {
            title: props.title,
            description: props.description,
            start: props.start && _jsx(EaselSurfaceHeaderBackButton, {
                ariaLabel: props.start.ariaLabel,
                onClick: props.start.onClick
            }),
            divider: props.divider ?? true,
            background: props.background ?? 'surface',
            end: props.end,
            padding: legacyPadding
        })
    });
}
