// @canva/app-ui-kit/dist/esm/ui/base/surface/internal/content_container.js
import { jsx as _jsx } from "react/jsx-runtime";
import 'react';
import { FocusLock } from '../../focus/focus_lock/focus_lock';
export function ContentContainer({ ariaLabel, ariaLabelledBy, ariaDescribedBy, captureFocus = true, autoFocus = true, role, onRequestEscape, children, ref, testId, className, id, onScroll, style }) {
    const divProps = {
        ['aria-label']: ariaLabel,
        ['aria-labelledby']: ariaLabelledBy,
        ['aria-describedby']: ariaDescribedBy,
        ['data-testid']: testId,
        role,
        id,
        onScroll,
        style
    };
    const effectiveCaptureFocus = role === 'dialog' || role === 'alertdialog' ? true : captureFocus;
    return _jsx(FocusLock, {
        returnFocus: effectiveCaptureFocus,
        preventScroll: "on-autofocus",
        disabled: !effectiveCaptureFocus,
        onRequestEscape: onRequestEscape,
        autoFocus: autoFocus,
        ref: ref,
        lockProps: divProps,
        className: className,
        makeContainerFocusable: true,
        children: children
    });
}
