// @canva/app-ui-kit/dist/cjs/ui/base/placeholder/private/placeholder.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get Placeholder () {
        return Placeholder;
    },
    get placeholderAnimationTypes () {
        return placeholderAnimationTypes;
    },
    get placeholderTransitionTypes () {
        return placeholderTransitionTypes;
    },
    get sizes () {
        return sizes;
    },
    get usePlaceholder () {
        return usePlaceholder;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _preconditions = require('../../../../base/preconditions');
const _classnames = _interop_require_default(require("classnames"));
const _react = _interop_require_wildcard(require("react"));
const _box = require('../../box/box');
const _provider = require('../../provider/provider');
const _ease = require('../../tokens/primitive/ease');
const _placeholdercss = _interop_require_wildcard(require("./placeholder.css"));
const _placeholdertestids = require("./placeholder.testids");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const TRANSITION_DURATION_SLOW_MS = 700;
const TRANSITION_DURATION_FAST_MS = 100;
const TRANSITION_TIMING_FUNCTION = _ease.easeInOut;
const placeholderAnimationTypes = [
    'none',
    'delayed',
    'immediate'
];
const placeholderTransitionTypes = [
    'none',
    'fast',
    'slow'
];
const sizes = [
    'fit-content',
    'full'
];
function usePlaceholder({ loading, animation: animationProp = 'delayed', transition: transitionProp = 'fast', animationIndex = 0, borderRadius = 'none' }) {
    const enableAnimations = (0, _provider.useEnableAnimations)();
    const animation = enableAnimations ? animationProp : 'none';
    const transition = enableAnimations ? transitionProp : 'none';
    const durationMs = transitionDurationMs(transition);
    const placeholderState = usePlaceholderTransitionState(loading, durationMs);
    const contentState = invertPlaceholderTransitionState(placeholderState);
    let showLoadingStyles;
    switch(placeholderState){
        case 'appearing':
        case 'visible':
            showLoadingStyles = true;
            break;
        case 'disappearing':
        case 'hidden':
            showLoadingStyles = false;
            break;
        default:
            throw new _preconditions.UnreachableError(placeholderState);
    }
    return {
        containerClassName: showLoadingStyles ? _placeholdercss.default.loading : _placeholdercss.default.loaded,
        containerCustomProperties: {
            [_placeholdercss.customProperties.transitionDuration]: `${durationMs.toFixed()}ms`,
            [_placeholdercss.customProperties.transitionTimingFunction]: TRANSITION_TIMING_FUNCTION
        },
        contentClassName: _placeholdercss.default.content,
        contentAriaHidden: contentState === 'hidden',
        renderVisual () {
            return (0, _jsxruntime.jsx)(_box.Box, {
                position: "absolute",
                inset: "0",
                ariaHidden: true,
                className: _placeholdercss.default.placeholderVisualContainer,
                children: (0, _jsxruntime.jsx)(_box.Box, {
                    position: "absolute",
                    inset: "0",
                    borderRadius: borderRadius,
                    className: (0, _classnames.default)(_placeholdercss.default.placeholderVisual, animationClassName(animation)),
                    customProperties: {
                        [_placeholdercss.customProperties.animationIndex]: animationIndex
                    },
                    testId: _placeholdertestids.PlaceholderTestIds.visual
                })
            });
        },
        transitionInfo: {
            placeholderState,
            durationMs,
            cssTimingFunction: _ease.easeInOut
        }
    };
}
function Placeholder({ loading, animation, transition, animationIndex, borderRadius, size: sizeProp, width: widthProp, height: heightProp, children }) {
    const { containerClassName, containerCustomProperties, contentClassName, contentAriaHidden, renderVisual } = usePlaceholder({
        loading,
        animation,
        transition,
        animationIndex,
        borderRadius
    });
    const width = widthProp ?? sizeProp ?? 'fit-content';
    const height = heightProp ?? sizeProp ?? 'fit-content';
    return (0, _jsxruntime.jsxs)(_box.Box, {
        className: (0, _classnames.default)(containerClassName, widthClassName(width), heightClassName(height)),
        customProperties: containerCustomProperties,
        position: "relative",
        testId: _placeholdertestids.PlaceholderTestIds.rootContainer,
        children: [
            (0, _jsxruntime.jsx)(_box.Box, {
                className: contentClassName,
                ariaHidden: contentAriaHidden,
                position: "relative",
                testId: _placeholdertestids.PlaceholderTestIds.childrenContainer,
                children: children
            }),
            renderVisual()
        ]
    });
}
function widthClassName(size) {
    switch(size){
        case 'fit-content':
            return _placeholdercss.default.widthFitContent;
        case 'full':
            return _placeholdercss.default.widthFull;
        default:
            throw new _preconditions.UnreachableError(size);
    }
}
function heightClassName(size) {
    switch(size){
        case 'fit-content':
            return _placeholdercss.default.heightFitContent;
        case 'full':
            return _placeholdercss.default.heightFull;
        default:
            throw new _preconditions.UnreachableError(size);
    }
}
function animationClassName(animation) {
    switch(animation){
        case 'none':
            return undefined;
        case 'delayed':
            return _placeholdercss.default.animationDelayed;
        case 'immediate':
            return _placeholdercss.default.animationImmediate;
        default:
            throw new _preconditions.UnreachableError(animation);
    }
}
function transitionDurationMs(transition) {
    switch(transition){
        case 'none':
            return 0;
        case 'fast':
            return TRANSITION_DURATION_FAST_MS;
        case 'slow':
            return TRANSITION_DURATION_SLOW_MS;
        default:
            throw new _preconditions.UnreachableError(transition);
    }
}
function usePlaceholderTransitionState(loading, durationMs) {
    const [transitionState, setTransitionState] = _react.useState(loading ? 'visible' : 'hidden');
    const timeoutRef = _react.useRef(undefined);
    const nextTransitionState = nextPlaceholderTransitionState(transitionState, loading, durationMs);
    _react.useEffect(()=>{
        if (nextTransitionState === transitionState) return;
        window.clearTimeout(timeoutRef.current);
        setTransitionState(nextTransitionState);
        if (nextTransitionState !== 'appearing' && nextTransitionState !== 'disappearing') return;
        const finalState = nextTransitionState === 'appearing' ? 'visible' : 'hidden';
        timeoutRef.current = window.setTimeout(()=>{
            timeoutRef.current = undefined;
            setTransitionState(finalState);
        }, durationMs);
    }, [
        durationMs,
        nextTransitionState,
        transitionState
    ]);
    _react.useEffect(()=>{
        return ()=>window.clearTimeout(timeoutRef.current);
    }, []);
    return nextTransitionState;
}
function nextPlaceholderTransitionState(transitionState, loading, durationMs) {
    if (durationMs === 0) return loading ? 'visible' : 'hidden';
    switch(transitionState){
        case 'visible':
            return loading ? 'visible' : 'disappearing';
        case 'hidden':
            return loading ? 'appearing' : 'hidden';
        case 'appearing':
        case 'disappearing':
            return loading ? 'appearing' : 'disappearing';
        default:
            throw new _preconditions.UnreachableError(transitionState);
    }
}
function invertPlaceholderTransitionState(state) {
    switch(state){
        case 'appearing':
            return 'disappearing';
        case 'visible':
            return 'hidden';
        case 'disappearing':
            return 'appearing';
        case 'hidden':
            return 'visible';
        default:
            throw new _preconditions.UnreachableError(state);
    }
}
