// @canva/app-ui-kit/dist/cjs/ui/base/responsive/private/responsive.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get conditions () {
        return conditions;
    },
    get generateStyle () {
        return generateStyle;
    },
    get mapResponsiveValueToStyleX () {
        return mapResponsiveValueToStyleX;
    },
    get resolveResponsiveObject () {
        return resolveResponsiveObject;
    }
});
const _breakpoints = require("../breakpoints");
const conditions = [
    'default',
    'smallUp',
    'mediumUp',
    'largeUp',
    'xLargeUp'
];
function normalizeResponsiveValue(value) {
    if (typeof value === 'string' || typeof value === 'number' || typeof value === 'boolean') return {
        default: value
    };
    return value;
}
function generateStyle(propertyMap, responsiveValue, getCssValue) {
    const normalizedValue = normalizeResponsiveValue(responsiveValue);
    return Object.fromEntries(Object.entries(normalizedValue).map(([condition, value])=>[
            propertyMap[condition],
            getCssValue ? getCssValue(value) : value
        ]));
}
function resolveResponsiveObject(value, breakpoint) {
    let resolvedValue;
    switch (breakpoint) {
case _breakpoints.Breakpoint.XLARGE:
    resolvedValue ?? (resolvedValue = value.xLargeUp);
case _breakpoints.Breakpoint.LARGE:
    resolvedValue ?? (resolvedValue = value.largeUp);
case _breakpoints.Breakpoint.MEDIUM:
    resolvedValue ?? (resolvedValue = value.mediumUp);
case _breakpoints.Breakpoint.SMALL:
    resolvedValue ?? (resolvedValue = value.smallUp);
case _breakpoints.Breakpoint.DEFAULT:
default:
    return resolvedValue ?? value.default;
} }
function mapResponsiveValueToStyleX(value, mapper, singleDynamicStyle, responsiveDynamicStyle) {
    if (value === undefined) return undefined;
    if (value == null || typeof value !== 'object') return singleDynamicStyle(mapper(value));
    const { default: defaultVal, smallUp = defaultVal, mediumUp = smallUp, largeUp = mediumUp, xLargeUp = largeUp } = value;
    return responsiveDynamicStyle({
        default: mapper(defaultVal),
        smallUp: mapper(smallUp),
        mediumUp: mapper(mediumUp),
        largeUp: mapper(largeUp),
        xLargeUp: mapper(xLargeUp)
    });
}
