// @canva/app-ui-kit/dist/esm/ui/base/shortcut/private/shortcut.js
import { jsx as _jsx } from "react/jsx-runtime";
import { composeLegacyClassNamePropWithStyles } from '../../../../platforms/libs/stylex/legacy_class_name';
import * as stylex from '../../../../platforms/libs/stylex/stylex';
import * as React from 'react';
import { makeHumanReadableArray, makeHumanReadableString } from '../../key_combinations/key_combinations';
import { useDirection } from '../../provider/provider';
import { easelPadding } from '../../style_helpers/easel_padding';
import { typography } from '../../tokens/composite_tokens';
import '../../tokens/primitive_tokens.stylex';
import { primitiveTokens } from '../../tokens/primitive_tokens.stylex';
import '../../tokens/semantic_tokens.stylex';
import { ShortcutTestIds } from './shortcut.testids';
export const Shortcut = React.memo(
    function Shortcut({ combination, platform, compact = false, size = 'medium', className }) {
        const direction = useDirection();
        const baseStyles = [
            typography.bodySmall,
            styles.key,
            size === 'small' && [
                styles.small,
                easelPadding.inline['0.5u']
            ],
            size === 'medium' && [
                styles.medium,
                easelPadding.inline['1u']
            ]
        ];
        const squareStyles = [
            styles.square,
            size === 'small' && styles.widthSmall,
            size === 'medium' && styles.widthMedium
        ];
        if (compact) {
            const str = makeHumanReadableString(combination, direction, platform);
            return _jsx("kbd", {
                "data-testid": ShortcutTestIds.key,
                ...composeLegacyClassNamePropWithStyles(className, baseStyles, str.length === 1 && squareStyles),
                children: str
            });
        }
        const keyStrings = makeHumanReadableArray(combination, direction, platform);
        if (keyStrings.length === 1) return _jsx("kbd", {
            "data-testid": ShortcutTestIds.key,
            ...composeLegacyClassNamePropWithStyles(className, baseStyles, keyStrings[0].length === 1 && squareStyles),
            children: keyStrings[0]
        });
        return _jsx("kbd", {
            ...composeLegacyClassNamePropWithStyles(className, styles.group),
            children: keyStrings.map((keyString)=>{
                return _jsx("kbd", {
                    "data-testid": ShortcutTestIds.key,
                    ...stylex.props([
                        baseStyles,
                        keyString.length === 1 && squareStyles,
                        styles.keyInGroup
                    ]),
                    children: keyString
                }, keyString);
            })
        });
    }
);
Shortcut.displayName = 'Shortcut';
primitiveTokens.baseUnitInRem;
primitiveTokens.baseUnitInRem;
const styles = {
    key: {
        kaIpWk: "xbssupt",
        kGNEyG: "x6s0dn4",
        kWkggS: "x17qqs1y",
        kB7OPa: "x9f619",
        kMwMTN: "x1p5nt2e",
        k1xSpc: "x1rg5ohu",
        ["$$css"]: true
    },
    square: {
        k8WAf4: "xt970qd",
        kg3NbH: "xnjsko4",
        k9WMMc: "x2b8uid",
        ["$$css"]: true
    },
    group: {
        ksoyAI: "xzt5al7",
        k1xSpc: "x3nfvp2",
        ["$$css"]: true
    },
    keyInGroup: {
        k71WvV: "x11k6k3q x1j53mea",
        keTefX: "x1wna4di x1pwwqoy",
        ["$$css"]: true
    },
    small: {
        kLWn49: "x1ym8nqh",
        kZKoxP: "x1teuabn",
        ["$$css"]: true
    },
    widthSmall: {
        kzqmXN: "x1o8o646",
        ["$$css"]: true
    },
    medium: {
        kLWn49: "x1rwdj4k",
        kZKoxP: "xxh6ah7",
        ["$$css"]: true
    },
    widthMedium: {
        kzqmXN: "x1s2tfyv",
        ["$$css"]: true
    }
};
