// @canva/app-ui-kit/dist/cjs/ui/base/metrics/metrics.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get Breakpoint () {
        return _breakpoints.Breakpoint;
    },
    get DEFAULT_ROOT_FONT_SIZE () {
        return DEFAULT_ROOT_FONT_SIZE;
    },
    get aboveLarge () {
        return _breakpoints.aboveLarge;
    },
    get aboveMedium () {
        return _breakpoints.aboveMedium;
    },
    get aboveSmall () {
        return _breakpoints.aboveSmall;
    },
    get aboveXLarge () {
        return _breakpoints.aboveXLarge;
    },
    get baseUnit () {
        return baseUnit;
    },
    get belowLarge () {
        return _breakpoints.belowLarge;
    },
    get belowMedium () {
        return _breakpoints.belowMedium;
    },
    get belowSmall () {
        return _breakpoints.belowSmall;
    },
    get belowXLarge () {
        return _breakpoints.belowXLarge;
    },
    get breakpointFromScreenWidth () {
        return _breakpoints.breakpointFromScreenWidth;
    },
    get getRootFontSize () {
        return getRootFontSize;
    },
    get getSpaceValue () {
        return getSpaceValue;
    },
    get largeBreakpoint () {
        return _breakpoints.largeBreakpoint;
    },
    get mediumBreakpoint () {
        return _breakpoints.mediumBreakpoint;
    },
    get resolveUnitOrNumberSize () {
        return resolveUnitOrNumberSize;
    },
    get smallBreakpoint () {
        return _breakpoints.smallBreakpoint;
    },
    get unitSizeToNumber () {
        return unitSizeToNumber;
    },
    get unitSpaces () {
        return unitSpaces;
    },
    get useBaseUnitInRem () {
        return useBaseUnitInRem;
    },
    get useTextZoomBaseUnit () {
        return useTextZoomBaseUnit;
    },
    get xLargeBreakpoint () {
        return _breakpoints.xLargeBreakpoint;
    }
});
const _provider = require('../provider/provider');
const _metricscss = _interop_require_default(require("./private/metrics.css"));
const _breakpoints = require('../responsive/breakpoints');
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const baseUnit = parseInt(_metricscss.default.baseUnit, 10) || 8;
const unitSpaces = [
    '0',
    '0.25u',
    '0.5u',
    '0.75u',
    '1u',
    '1.25u',
    '1.5u',
    '2u',
    '3u',
    '4u',
    '6u',
    '8u',
    '12u'
];
const getSpaceValue = (space)=>{
    return spaceValues[space];
};
const spaceValues = {
    ['0']: _metricscss.default.space0,
    ['0.25u']: _metricscss.default.space025,
    ['0.5u']: _metricscss.default.space050,
    ['0.75u']: _metricscss.default.space075,
    ['1u']: _metricscss.default.space1,
    ['1.25u']: _metricscss.default.space125,
    ['1.5u']: _metricscss.default.space150,
    ['2u']: _metricscss.default.space2,
    ['3u']: _metricscss.default.space3,
    ['4u']: _metricscss.default.space4,
    ['6u']: _metricscss.default.space6,
    ['8u']: _metricscss.default.space8,
    ['12u']: _metricscss.default.space12
};
const unitSizeToNumber = (unitSize)=>{
    return parseFloat(unitSize) * baseUnit;
};
const resolveUnitOrNumberSize = (size)=>{
    if (size == null || typeof size === 'number') return size;
    return unitSizeToNumber(size);
};
const DEFAULT_ROOT_FONT_SIZE = 10;
const useBaseUnitInRem = ()=>{
    const rootFontSize = (0, _provider.useExpectedPixelsPerRem)() || DEFAULT_ROOT_FONT_SIZE;
    return baseUnit / rootFontSize;
};
const getRootFontSize = ()=>{
    if (window == null) return DEFAULT_ROOT_FONT_SIZE;
    return parseFloat(getComputedStyle(window.document.documentElement).fontSize) || DEFAULT_ROOT_FONT_SIZE;
};
const useTextZoomBaseUnit = (currentRootFontSize = getRootFontSize())=>{
    const originalRootFontSize = (0, _provider.useExpectedPixelsPerRem)() || DEFAULT_ROOT_FONT_SIZE;
    const textZoomLevel = currentRootFontSize / originalRootFontSize;
    return baseUnit * textZoomLevel;
};
