// @canva/app-ui-kit/dist/cjs/ui/base/shortcut/private/shortcut.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "Shortcut", {
    enumerable: true,
    get: function() {
        return Shortcut;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _legacy_class_name = require('../../../../platforms/libs/stylex/legacy_class_name');
const _stylex = _interop_require_wildcard(require('../../../../platforms/libs/stylex/stylex'));
const _react = _interop_require_wildcard(require("react"));
const _key_combinations = require('../../key_combinations/key_combinations');
const _provider = require('../../provider/provider');
const _easel_padding = require('../../style_helpers/easel_padding');
const _composite_tokens = require('../../tokens/composite_tokens');
const _primitive_tokensstylex = require('../../tokens/primitive_tokens.stylex');
require('../../tokens/semantic_tokens.stylex');
const _shortcuttestids = require("./shortcut.testids");
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const Shortcut = _react.memo(
    function Shortcut({ combination, platform, compact = false, size = 'medium', className }) {
        const direction = (0, _provider.useDirection)();
        const baseStyles = [
            _composite_tokens.typography.bodySmall,
            styles.key,
            size === 'small' && [
                styles.small,
                _easel_padding.easelPadding.inline['0.5u']
            ],
            size === 'medium' && [
                styles.medium,
                _easel_padding.easelPadding.inline['1u']
            ]
        ];
        const squareStyles = [
            styles.square,
            size === 'small' && styles.widthSmall,
            size === 'medium' && styles.widthMedium
        ];
        if (compact) {
            const str = (0, _key_combinations.makeHumanReadableString)(combination, direction, platform);
            return (0, _jsxruntime.jsx)("kbd", {
                "data-testid": _shortcuttestids.ShortcutTestIds.key,
                ...(0, _legacy_class_name.composeLegacyClassNamePropWithStyles)(className, baseStyles, str.length === 1 && squareStyles),
                children: str
            });
        }
        const keyStrings = (0, _key_combinations.makeHumanReadableArray)(combination, direction, platform);
        if (keyStrings.length === 1) return (0, _jsxruntime.jsx)("kbd", {
            "data-testid": _shortcuttestids.ShortcutTestIds.key,
            ...(0, _legacy_class_name.composeLegacyClassNamePropWithStyles)(className, baseStyles, keyStrings[0].length === 1 && squareStyles),
            children: keyStrings[0]
        });
        return (0, _jsxruntime.jsx)("kbd", {
            ...(0, _legacy_class_name.composeLegacyClassNamePropWithStyles)(className, styles.group),
            children: keyStrings.map((keyString)=>{
                return (0, _jsxruntime.jsx)("kbd", {
                    "data-testid": _shortcuttestids.ShortcutTestIds.key,
                    ..._stylex.props([
                        baseStyles,
                        keyString.length === 1 && squareStyles,
                        styles.keyInGroup
                    ]),
                    children: keyString
                }, keyString);
            })
        });
    }
);
Shortcut.displayName = 'Shortcut';
_primitive_tokensstylex.primitiveTokens.baseUnitInRem;
_primitive_tokensstylex.primitiveTokens.baseUnitInRem;
const styles = {
    key: {
        kaIpWk: "xbssupt",
        kGNEyG: "x6s0dn4",
        kWkggS: "x17qqs1y",
        kB7OPa: "x9f619",
        kMwMTN: "x1p5nt2e",
        k1xSpc: "x1rg5ohu",
        ["$$css"]: true
    },
    square: {
        k8WAf4: "xt970qd",
        kg3NbH: "xnjsko4",
        k9WMMc: "x2b8uid",
        ["$$css"]: true
    },
    group: {
        ksoyAI: "xzt5al7",
        k1xSpc: "x3nfvp2",
        ["$$css"]: true
    },
    keyInGroup: {
        k71WvV: "x11k6k3q x1j53mea",
        keTefX: "x1wna4di x1pwwqoy",
        ["$$css"]: true
    },
    small: {
        kLWn49: "x1ym8nqh",
        kZKoxP: "x1teuabn",
        ["$$css"]: true
    },
    widthSmall: {
        kzqmXN: "x1o8o646",
        ["$$css"]: true
    },
    medium: {
        kLWn49: "x1rwdj4k",
        kZKoxP: "xxh6ah7",
        ["$$css"]: true
    },
    widthMedium: {
        kzqmXN: "x1s2tfyv",
        ["$$css"]: true
    }
};
