// @canva/app-ui-kit/dist/cjs/ui/base/box/private/box.js
"use strict"
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get Box () {
        return Box;
    },
    get useBox () {
        return useBox;
    }
});
const _jsxruntime = require("react/jsx-runtime");
const _stylex = _interop_require_wildcard(require('../../../../platforms/libs/stylex/stylex'));
const _callback_ref = require('../../../../base/react/callback_ref');
const _classnames = _interop_require_default(require("classnames"));
require("react");
const _common_props = require('../common_props/common_props');
const _checkered_patterncss = _interop_require_default(require('../../checkered_pattern/checkered_pattern.css'));
const _box_styles = require("./box_styles");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
function useBox({ className: classNameProp, customProperties, sx, ...options }) {
    const boxProps = _stylex.props([
        (0, _box_styles.useBoxStyles)(options),
        sx
    ]);
    return {
        className: (0, _classnames.default)(boxProps.className, classNameProp, getCheckeredPatternClassName(options)),
        style: {
            ...boxProps.style,
            ...customProperties
        },
        ['data-style-src']: boxProps['data-style-src']
    };
}
function Box(props) {
    const { tagName: TagName = 'div', children, tabIndex, ref, ...rest } = props;
    return (0, _jsxruntime.jsx)(TagName, {
        ref: (0, _callback_ref.toCallbackRef)(ref),
        ...useBox(rest),
        ...(0, _common_props.commonPropsToDOMProps)(rest),
        tabIndex: tabIndex,
        children: children
    });
}
function getCheckeredPatternClassName({ background }) {
    return background === 'checkered' ? _checkered_patterncss.default.checkeredPattern : null;
}
