// @canva/app-ui-kit/dist/esm/ui/base/thumbnail/private/thumbnail.js
import { jsx as _jsx } from "react/jsx-runtime";
import { composeLegacyClassNamePropWithStyles } from '../../../../platforms/libs/stylex/legacy_class_name';
import * as React from 'react';
import { resolveUnitOrNumberSize } from '../../metrics/metrics';
import { getDimensions } from './dimension';
import { resolveAspectRatio, SizingContainer, ThumbnailContainerContext } from './thumbnail_container';
import { sharedThumbnailStyles } from './thumbnail_styles';
export const Thumbnail = ({ width, height, aspectRatio, border = 'low', borderRadius = 'elementSmall', children, innerContainerClassName })=>{
    const thumbnailContainerContext = React.useContext(ThumbnailContainerContext);
    const thumbnailRatio = resolveAspectRatio({
        width,
        height,
        aspectRatio
    });
    const { aspectRatio: containerRatio, padding, fit } = thumbnailContainerContext || {};
    const needDimensions = thumbnailRatio != null && containerRatio != null;
    if (!needDimensions && border === 'none' && borderRadius === 'none' && innerContainerClassName == null)
    return _jsx(SizingContainer, {
        height: height,
        width: width,
        aspectRatio: thumbnailRatio,
        children: children
    });
    const resolvedPadding = padding === 'none' ? 0 : resolveUnitOrNumberSize(padding);
    const dimensionStyles = needDimensions ? getDimensions(fit, containerRatio, thumbnailRatio, resolvedPadding) : undefined;
    return _jsx(SizingContainer, {
        height: height,
        width: width,
        aspectRatio: needDimensions ? undefined : thumbnailRatio,
        children: _jsx("div", {
            ...composeLegacyClassNamePropWithStyles(innerContainerClassName, [
                styles.root,
                border !== 'none' && sharedThumbnailStyles.border,
                border === 'low' && sharedThumbnailStyles.lowBorder,
                borderRadius === 'elementSmall' && sharedThumbnailStyles.elementSmallBorderRadius,
                borderRadius === 'element' && sharedThumbnailStyles.elementBorderRadius,
                borderRadius === 'elementRelaxed' && sharedThumbnailStyles.elementRelaxedBorderRadius,
                borderRadius === 'elementSoftest' && sharedThumbnailStyles.elementSoftestBorderRadius,
                borderRadius === 'elementRound' && sharedThumbnailStyles.elementRoundBorderRadius
            ]),
            style: dimensionStyles,
            children: children
        })
    });
};
const styles = {
    root: {
        k1xSpc: "x78zum5",
        kZKoxP: "x5yr21d",
        kzqmXN: "xh8yej3",
        ["$$css"]: true
    }
};
