// @canva/app-ui-kit/dist/esm/ui/base/thumbnail/private/thumbnail_container.js
import { jsx as _jsx } from "react/jsx-runtime";
import { composeLegacyClassNamePropWithStyles } from '../../../../platforms/libs/stylex/legacy_class_name';
import * as React from 'react';
import { AspectRatio } from '../../aspect_ratio/aspect_ratio';
import { resolveUnitOrNumberSize } from '../../metrics/metrics';
import '../../tokens/primitive_tokens.stylex';
import '../../tokens/semantic_tokens.stylex';
import { sharedThumbnailStyles } from './thumbnail_styles';
export const ThumbnailContainerContext = React.createContext(undefined);
export const useInThumbnailContainerContext = ()=>{
    const containerContext = React.useContext(ThumbnailContainerContext);
    return containerContext != null;
};
export const ThumbnailContainer = ({ children, className, background = 'secondary', border = 'none', borderRadius = 'elementSmall', fit = 'cover', padding = '2u', height, width, aspectRatio: aspectRatioProp })=>{
    const aspectRatio = resolveAspectRatio({
        width,
        height,
        aspectRatio: aspectRatioProp
    });
    return _jsx(SizingContainer, {
        height: height,
        width: width,
        aspectRatio: aspectRatio,
        children: _jsx("div", {
            ...composeLegacyClassNamePropWithStyles(className, [
                styles.root,
                ...getThumbnailContainerStyles({
                    background,
                    border,
                    borderRadius,
                    padding
                })
            ]),
            children: _jsx(ThumbnailContainerContext, {
                value: {
                    aspectRatio,
                    padding,
                    fit
                },
                children: children
            })
        })
    });
};
const getThumbnailContainerStyles = ({ background, border, borderRadius, padding })=>[
        background === 'secondary' && styles.secondaryBackground,
        background === 'contrastOnLight' && styles.contrastOnLight,
        background === 'contrastOnDark' && styles.contrastOnDark,
        background === 'secondaryHover' && styles.secondaryBackgroundHover,
        background === 'contrastOnLightHover' && styles.contrastOnLightHover,
        background === 'contrastOnDarkHover' && styles.contrastOnDarkHover,
        border !== 'none' && sharedThumbnailStyles.border,
        border === 'low' && sharedThumbnailStyles.lowBorder,
        borderRadius === 'elementSmall' && sharedThumbnailStyles.elementSmallBorderRadius,
        borderRadius === 'element' && sharedThumbnailStyles.elementBorderRadius,
        borderRadius === 'elementRelaxed' && sharedThumbnailStyles.elementRelaxedBorderRadius,
        borderRadius === 'elementSoftest' && sharedThumbnailStyles.elementSoftestBorderRadius,
        borderRadius === 'elementRound' && sharedThumbnailStyles.elementRoundBorderRadius,
        padding === '1u' && styles.paddingOneUnit,
        padding === '2u' && styles.paddingTwoUnit
    ];
export const SizingContainer = ({ aspectRatio, width, height, children })=>{
    const widthInPixel = resolveUnitOrNumberSize(width);
    const heightInPixel = resolveUnitOrNumberSize(height);
    if (aspectRatio != null) return _jsx(AspectRatio, {
        ratio: aspectRatio,
        width: width,
        height: height,
        children: children
    });
    if (width != null || height != null) return _jsx("div", {
        style: {
            width: widthInPixel ?? '100%',
            height: heightInPixel ?? '100%',
            flexShrink: 0
        },
        children: children
    });
    return children;
};
export const resolveAspectRatio = ({ aspectRatio, width, height })=>{
    const widthInPixel = resolveUnitOrNumberSize(width);
    const heightInPixel = resolveUnitOrNumberSize(height);
    return aspectRatio ?? (widthInPixel != null && heightInPixel != null ? widthInPixel / heightInPixel : undefined);
};
const styles = {
    root: {
        kVQacm: "xb3r6kr",
        kGNEyG: "x6s0dn4",
        k1xSpc: "x78zum5",
        kjj79g: "xl56j7k",
        kZKoxP: "x5yr21d",
        kzqmXN: "xh8yej3",
        ["$$css"]: true
    },
    secondaryBackground: {
        kWkggS: "x1gim84k",
        ["$$css"]: true
    },
    secondaryBackgroundHover: {
        kWkggS: "xbdlxvi",
        ["$$css"]: true
    },
    contrastOnLight: {
        kWkggS: "x110c1mw",
        ["$$css"]: true
    },
    contrastOnLightHover: {
        kWkggS: "xqxvuf3",
        ["$$css"]: true
    },
    contrastOnDark: {
        kWkggS: "x9eh2el",
        ["$$css"]: true
    },
    contrastOnDarkHover: {
        kWkggS: "xong8os",
        ["$$css"]: true
    },
    paddingOneUnit: {
        kmVPX3: "xmom2u2",
        kB7OPa: "x9f619",
        ["$$css"]: true
    },
    paddingTwoUnit: {
        kmVPX3: "x5vm2tf",
        kB7OPa: "x9f619",
        ["$$css"]: true
    }
};
