// @canva/app-ui-kit/dist/esm/ui/base/placeholder/private/placeholder.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { UnreachableError } from '../../../../base/preconditions';
import classNames from 'classnames';
import * as React from 'react';
import { Box } from '../../box/box';
import { useEnableAnimations } from '../../provider/provider';
import { easeInOut } from '../../tokens/primitive/ease';
import styles, { customProperties } from './placeholder.css';
import { PlaceholderTestIds } from './placeholder.testids';
const TRANSITION_DURATION_SLOW_MS = 700;
const TRANSITION_DURATION_FAST_MS = 100;
const TRANSITION_TIMING_FUNCTION = easeInOut;
export const placeholderAnimationTypes = [
    'none',
    'delayed',
    'immediate'
];
export const placeholderTransitionTypes = [
    'none',
    'fast',
    'slow'
];
export const sizes = [
    'fit-content',
    'full'
];
export function usePlaceholder(
    { loading, animation: animationProp = 'delayed', transition: transitionProp = 'fast', animationIndex = 0, borderRadius = 'none' }
) {
    const enableAnimations = useEnableAnimations();
    const animation = enableAnimations ? animationProp : 'none';
    const transition = enableAnimations ? transitionProp : 'none';
    const durationMs = transitionDurationMs(transition);
    const placeholderState = usePlaceholderTransitionState(loading, durationMs);
    const contentState = invertPlaceholderTransitionState(placeholderState);
    let showLoadingStyles;
    switch(placeholderState){
        case 'appearing':
        case 'visible':
            showLoadingStyles = true;
            break;
        case 'disappearing':
        case 'hidden':
            showLoadingStyles = false;
            break;
        default:
            throw new UnreachableError(placeholderState);
    }
    return {
        containerClassName: showLoadingStyles ? styles.loading : styles.loaded,
        containerCustomProperties: {
            [customProperties.transitionDuration]: `${durationMs.toFixed()}ms`,
            [customProperties.transitionTimingFunction]: TRANSITION_TIMING_FUNCTION
        },
        contentClassName: styles.content,
        contentAriaHidden: contentState === 'hidden',
        renderVisual () {
            return _jsx(Box, {
                position: "absolute",
                inset: "0",
                ariaHidden: true,
                className: styles.placeholderVisualContainer,
                children: _jsx(Box, {
                    position: "absolute",
                    inset: "0",
                    borderRadius: borderRadius,
                    className: classNames(styles.placeholderVisual, animationClassName(animation)),
                    customProperties: {
                        [customProperties.animationIndex]: animationIndex
                    },
                    testId: PlaceholderTestIds.visual
                })
            });
        },
        transitionInfo: {
            placeholderState,
            durationMs,
            cssTimingFunction: easeInOut
        }
    };
}
export function Placeholder(
    { loading, animation, transition, animationIndex, borderRadius, size: sizeProp, width: widthProp, height: heightProp, children }
) {
    const { containerClassName, containerCustomProperties, contentClassName, contentAriaHidden, renderVisual } = usePlaceholder({
        loading,
        animation,
        transition,
        animationIndex,
        borderRadius
    });
    const width = widthProp ?? sizeProp ?? 'fit-content';
    const height = heightProp ?? sizeProp ?? 'fit-content';
    return _jsxs(Box, {
        className: classNames(containerClassName, widthClassName(width), heightClassName(height)),
        customProperties: containerCustomProperties,
        position: "relative",
        testId: PlaceholderTestIds.rootContainer,
        children: [
            _jsx(Box, {
                className: contentClassName,
                ariaHidden: contentAriaHidden,
                position: "relative",
                testId: PlaceholderTestIds.childrenContainer,
                children: children
            }),
            renderVisual()
        ]
    });
}
function widthClassName(size) {
    switch(size){
        case 'fit-content':
            return styles.widthFitContent;
        case 'full':
            return styles.widthFull;
        default:
            throw new UnreachableError(size);
    }
}
function heightClassName(size) {
    switch(size){
        case 'fit-content':
            return styles.heightFitContent;
        case 'full':
            return styles.heightFull;
        default:
            throw new UnreachableError(size);
    }
}
function animationClassName(animation) {
    switch(animation){
        case 'none':
            return undefined;
        case 'delayed':
            return styles.animationDelayed;
        case 'immediate':
            return styles.animationImmediate;
        default:
            throw new UnreachableError(animation);
    }
}
function transitionDurationMs(transition) {
    switch(transition){
        case 'none':
            return 0;
        case 'fast':
            return TRANSITION_DURATION_FAST_MS;
        case 'slow':
            return TRANSITION_DURATION_SLOW_MS;
        default:
            throw new UnreachableError(transition);
    }
}
function usePlaceholderTransitionState(loading, durationMs) {
    const [transitionState, setTransitionState] = React.useState(loading ? 'visible' : 'hidden');
    const timeoutRef = React.useRef(undefined);
    const nextTransitionState = nextPlaceholderTransitionState(transitionState, loading, durationMs);
    React.useEffect(()=>{
        if (nextTransitionState === transitionState) return;
        window.clearTimeout(timeoutRef.current);
        setTransitionState(nextTransitionState);
        if (nextTransitionState !== 'appearing' && nextTransitionState !== 'disappearing') return;
        const finalState = nextTransitionState === 'appearing' ? 'visible' : 'hidden';
        timeoutRef.current = window.setTimeout(()=>{
            timeoutRef.current = undefined;
            setTransitionState(finalState);
        }, durationMs);
    }, [
        durationMs,
        nextTransitionState,
        transitionState
    ]);
    React.useEffect(()=>{
        return ()=>window.clearTimeout(timeoutRef.current);
    }, []);
    return nextTransitionState;
}
function nextPlaceholderTransitionState(transitionState, loading, durationMs) {
    if (durationMs === 0) return loading ? 'visible' : 'hidden';
    switch(transitionState){
        case 'visible':
            return loading ? 'visible' : 'disappearing';
        case 'hidden':
            return loading ? 'appearing' : 'hidden';
        case 'appearing':
        case 'disappearing':
            return loading ? 'appearing' : 'disappearing';
        default:
            throw new UnreachableError(transitionState);
    }
}
function invertPlaceholderTransitionState(state) {
    switch(state){
        case 'appearing':
            return 'disappearing';
        case 'visible':
            return 'hidden';
        case 'disappearing':
            return 'appearing';
        case 'hidden':
            return 'visible';
        default:
            throw new UnreachableError(state);
    }
}
